{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "32091830",
   "metadata": {},
   "source": [
    "## L10: Simple Pattern Matching\n",
    "Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf538e53",
   "metadata": {},
   "source": [
    "The first 4 exercises will all have to do with the files data1-4.gb, which are various Genbank entries of genes. First you should study the files, notice the structure of the data. In all exercises you will have to parse (read and find the wanted data) the files using pattern matching. Every exercise adds to the previous ones, so the final program can do a lot. Remember. Your program should be able to handle all files, but just one at a time.\n",
    "\n",
    "1\\. Extract the accession number, the definition and the organism (and print it)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "66483c38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a file Genbank entry file, such as one of data1-4.gb: data2.gb\n",
      "Specify what you want to know by typing one of: accession number, definition, organism (STOP to quit search): STOP\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "prompt user for file name\n",
    "with open file name:\n",
    "    for line in file\n",
    "        accession number can be found by \"ACCESSION\", assign to variable\n",
    "        definition can be fond by \"DEFINITION\", assign to variable\n",
    "        organism can be found by \"ORGANISM\", assign to variable\n",
    "display each variable for sanity check\n",
    "\"\"\"\n",
    "# code\n",
    "def main():\n",
    "    usr_filename = input(\"Please enter a file Genbank entry file, such as one of data1-4.gb: \")\n",
    "    find_ac_def_os(usr_filename)\n",
    "\n",
    "\n",
    "def find_ac_def_os(filename):\n",
    "    while True:\n",
    "        with open (f\"{filename}\", \"r\") as infile:\n",
    "            # make a loop so usr can keep prompting and we can keep displaying\n",
    "                search_word = input(\"Specify what you want to know by typing one of: accession number, definition, organism (STOP to quit search): \")\n",
    "                for line in infile:\n",
    "                    # line.strip() will help with key word search like ORGANISM line, which starts with two spaces.\n",
    "                    line = line.strip()\n",
    "                    # the find start parts can be made into a smaller programme, but not that fit for what ex1 asks for.\n",
    "                    # the way to do this is to ask the user for a string, that equals to \"ACCESSION\", \"DEFINITION\", ...etc\n",
    "                    # search_word.split()[0] is useful for cases when we write accession number.\n",
    "                    if line.startswith(f\"{search_word.split()[0].upper()}\"):\n",
    "                        result = line.strip(f\"{search_word.split()[0].upper()}\").strip()\n",
    "                        print(f\"The {search_word} is {result}.\")\n",
    "                    elif len(search_word) == 0:\n",
    "                        print(\"Empty search term!\")\n",
    "                    else:\n",
    "                        # using pass here for now\n",
    "                        pass\n",
    "                if search_word == \"STOP\":\n",
    "                    break\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6fda36e",
   "metadata": {},
   "source": [
    "2\\. Extract and print all MEDLINE article numbers which are mentioned in the entries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "43c677ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a file Genbank entry file, such as one of data1-4.gb: data3.gb\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline (STOP to quit search): organism\n",
      "The organism is Salmo salar.\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline (STOP to quit search): STOP\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "initiate an empty list to keep track of all MEDLINE values.\n",
    "\"\"\"\n",
    "# code\n",
    "def main():\n",
    "    usr_filename = input(\"Please enter a file Genbank entry file, such as one of data1-4.gb: \")\n",
    "    find_keyword(usr_filename)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def find_keyword(filename):\n",
    "# make a loop so usr can keep prompting and we can keep displaying\n",
    "    while True:\n",
    "        search_word = input(\"Specify what you want to know by typing one of the following:\\naccession number, definition, organism, medline (STOP to quit search): \")\n",
    "        if search_word == 'medline':\n",
    "            find_medline(filename)\n",
    "        elif search_word in (\"accession number\", \"definition\", \"organism\"):\n",
    "            find_an_df_og(filename, search_word)    # very important to include \"search_word\" here.\n",
    "        elif search_word == \"STOP\":\n",
    "            break\n",
    "        else:\n",
    "            print(\"I don't think we have what you typed, let's try again.\")\n",
    "            pass\n",
    "\n",
    "\n",
    "\n",
    "def find_medline(filename):\n",
    "    medlines = []\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"MEDLINE\"):\n",
    "                medlines.append(line.strip(\"MEDLINE\").strip())\n",
    "            else:\n",
    "                pass\n",
    "    if len(medlines) > 1:\n",
    "        are = \"are\"\n",
    "    else:\n",
    "        are = \"is\"\n",
    "    print(f\"The medlines {are}: {medlines}\")\n",
    "\n",
    "\n",
    "\n",
    "def find_an_df_og(filename, search_word):\n",
    "    with open (f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # line.strip() will help with key word search like ORGANISM line, which starts with two spaces.\n",
    "            line = line.strip()\n",
    "            # the find start parts can be made into a smaller programme, but not that fit for what ex1 asks for.\n",
    "            # the way to do this is to ask the user for a string, that equals to \"ACCESSION\", \"DEFINITION\", ...etc\n",
    "            # search_word.split()[0] is useful for cases when we write accession number.\n",
    "            if line.startswith(f\"{search_word.split()[0].upper()}\"):\n",
    "                result = line.strip(f\"{search_word.split()[0].upper()}\").strip()\n",
    "                print(f\"The {search_word} is {result}.\")\n",
    "            elif len(search_word) == 0:\n",
    "                print(\"Empty search term!\")\n",
    "            else:\n",
    "                # using pass here for now\n",
    "                pass\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "558c0ad8",
   "metadata": {},
   "source": [
    "3\\. Extract and print the translated gene (the amino acid sequence). Look for the line starting with /translation=. Generalize; An amino acid sequence can be short, i.e. only one line in the feature table, or long, i.e. more than one line in the feature table. Use stateful parsing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1e12aa7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a file Genbank entry file, such as one of data1-4.gb: data4.gb\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline, translated (STOP to quit search): translated\n",
      "\n",
      "The DNA sequence of the translated gene is:\n",
      "HLHQLAADTYKEFERAYIPEGQRYSIQNAQAAFCFSETIPAPTGKEEAQQRSDMELLRFSLLLIQSWLGPVQFLSRVFTNSLVFGTSDRVFEKLKDLEEGIQALMRELEDGSLRAGQLLKQTYDKFDTNMRSDDALLKNYGLLSCFKKDLHKAETYLRVMKCRRFVESSCAFg_peptide     join(145..154,357..424)mat_peptide     join(425..517,712..828,965..1126,1373..1570)product=\"growth hormone155..356umber=1\n",
      "\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline, translated (STOP to quit search): STOP\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "need to consider start with and end with\n",
    "spanning multiple lines, so can do some kind of Green Red thing!\n",
    "Go back to check\n",
    "flagging\n",
    "\"\"\"\n",
    "\n",
    "# code\n",
    "def main():\n",
    "    usr_filename = input(\"Please enter a file Genbank entry file, such as one of data1-4.gb: \")\n",
    "    find_keyword(usr_filename)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def find_keyword(filename):\n",
    "# make a loop so usr can keep prompting and we can keep displaying\n",
    "    while True:\n",
    "        search_word = input(\"Specify what you want to know by typing one of the following:\\naccession number, definition, organism, medline, translated (STOP to quit search): \")\n",
    "        if search_word == \"medline\":\n",
    "            find_medline(filename)\n",
    "        elif search_word in (\"accession number\", \"definition\", \"organism\"):\n",
    "            find_an_df_og(filename, search_word)    # very important to include \"search_word\" here.\n",
    "        elif search_word == \"translated\":\n",
    "            find_translated(filename)\n",
    "        elif search_word == \"STOP\":\n",
    "            break\n",
    "        else:\n",
    "            print(\"I don't think we have what you typed, let's try again.\")\n",
    "            pass\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def find_translated(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, tran_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            # this is the red line\n",
    "            # put the red line at the top so it is not included in the result :)\n",
    "            # this way the results stop at the correct place.\n",
    "            if line.startswith(\"exon\"):\n",
    "                tran_flag = False\n",
    "            if tran_flag:\n",
    "                # carful here, the keyword in the file is translation, not translated :).\n",
    "                result += line.strip(\"/translation=\").strip('\"').strip()\n",
    "            # this is the green line\n",
    "            # carful here with, remember the correct word is translation.\n",
    "            if line.startswith(\"/translation=\"):\n",
    "                tran_flag = True\n",
    "        if tran_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(f\"\\nThe AA sequence of the translated gene is:\\n{result}\\n\")\n",
    "\n",
    "\n",
    "def find_medline(filename):\n",
    "    medlines = []\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"MEDLINE\"):\n",
    "                medlines.append(line.strip(\"MEDLINE\").strip())\n",
    "            else:\n",
    "                pass\n",
    "    if len(medlines) > 1:\n",
    "        are = \"are\"\n",
    "    else:\n",
    "        are = \"is\"\n",
    "    print(f\"The medlines {are}: {medlines}\")\n",
    "\n",
    "\n",
    "def find_an_df_og(filename, search_word):\n",
    "    with open (f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # line.strip() will help with key word search like ORGANISM line, which starts with two spaces.\n",
    "            line = line.strip()\n",
    "            # the find start parts can be made into a smaller programme, but not that fit for what ex1 asks for.\n",
    "            # the way to do this is to ask the user for a string, that equals to \"ACCESSION\", \"DEFINITION\", ...etc\n",
    "            # search_word.split()[0] is useful for cases when we write accession number.\n",
    "            if line.startswith(f\"{search_word.split()[0].upper()}\"):\n",
    "                result = line.strip(f\"{search_word.split()[0].upper()}\").strip()\n",
    "                print(f\"The {search_word} is {result}.\")\n",
    "            elif len(search_word) == 0:\n",
    "                print(\"Empty search term!\")\n",
    "            else:\n",
    "                # using pass here for now\n",
    "                pass\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6845926",
   "metadata": {},
   "source": [
    "4\\. Extract and print the DNA (whole base sequence in the end of the file). Use stateful parsing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d6633edc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a file Genbank entry file, such as one of data1-4.gb: data1.gb\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline, translated, DNA (STOP to quit search): translated\n",
      "\n",
      "The DNA sequence of the translated gene is:\n",
      "ERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCNg_peptide     2424..2495gene=\"INSe=\"G00-119-349mat_peptide     join(2496..2585,2586..2610,3397..3476,3477..3539)gene=\"INSe=\"c peptide; G00-119-3492611..3396gene=\"INSe=\"G00-119-349umber=2variation       3229gene=\"INSe=\"c in alpha-allele; g in beta-allele ([4])\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "Interrupted by user",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/9c/bz3_nczs67l3rlf5shp54rjh0000gn/T/ipykernel_39995/3097521196.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    119\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    120\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 121\u001b[0;31m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/9c/bz3_nczs67l3rlf5shp54rjh0000gn/T/ipykernel_39995/3097521196.py\u001b[0m in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0musr_filename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Please enter a file Genbank entry file, such as one of data1-4.gb: \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m     \u001b[0mfind_keyword\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0musr_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/9c/bz3_nczs67l3rlf5shp54rjh0000gn/T/ipykernel_39995/3097521196.py\u001b[0m in \u001b[0;36mfind_keyword\u001b[0;34m(filename)\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;31m# make a loop so usr can keep prompting and we can keep displaying\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 29\u001b[0;31m         \u001b[0msearch_word\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Specify what you want to know by typing one of the following:\\naccession number, definition, organism, medline, translated, DNA (STOP to quit search): \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     30\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0msearch_word\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"medline\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m             \u001b[0mfind_medline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/anaconda3/lib/python3.9/site-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36mraw_input\u001b[0;34m(self, prompt)\u001b[0m\n\u001b[1;32m   1004\u001b[0m                 \u001b[0;34m\"raw_input was called, but this frontend does not support input requests.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1005\u001b[0m             )\n\u001b[0;32m-> 1006\u001b[0;31m         return self._input_request(\n\u001b[0m\u001b[1;32m   1007\u001b[0m             \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprompt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1008\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_ident\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"shell\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/anaconda3/lib/python3.9/site-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36m_input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m   1049\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1050\u001b[0m                 \u001b[0;31m# re-raise KeyboardInterrupt, to truncate traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1051\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Interrupted by user\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1052\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1053\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwarning\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid Message:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_info\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: Interrupted by user"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "reference what you did in ex3.py\n",
    "! line[:1] does not work that well, but doing it using startswith() works well, (see method used in ex3) :).\n",
    "flag is False when line[:-1] == //\n",
    "flag is True when line[:-1] == \"ORIGIN\"\n",
    "when appending the data we can do something like:\n",
    "initiate data as an empty string\n",
    "example line =\n",
    "\"       61 tgatcttttc cagagccaat cacatgcgcg cctccagtgg gcggagtgca ctcataaata\"\n",
    "for char in line.strip():\n",
    "    if char.isalpha():\n",
    "        data += char\n",
    "also tried using a string method, but still not working :<\n",
    "the bit within if flag:...i.e what it's supopse to do if the flag is True, gets skipped.\n",
    "changed approach from [:-1] to .startswith() and it worked\n",
    "\"\"\"\n",
    "# code\n",
    "def main():\n",
    "    usr_filename = input(\"Please enter a file Genbank entry file, such as one of data1-4.gb: \")\n",
    "    find_keyword(usr_filename)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def find_keyword(filename):\n",
    "# make a loop so usr can keep prompting and we can keep displaying\n",
    "    while True:\n",
    "        search_word = input(\"Specify what you want to know by typing one of the following:\\naccession number, definition, organism, medline, translated, DNA (STOP to quit search): \")\n",
    "        if search_word == \"medline\":\n",
    "            find_medline(filename)\n",
    "        elif search_word in (\"accession number\", \"definition\", \"organism\"):\n",
    "            find_an_df_og(filename, search_word)    # very important to include \"search_word\" here.\n",
    "        elif search_word == \"translated\":\n",
    "            find_translated(filename)\n",
    "        elif search_word == \"DNA\":\n",
    "            find_DNA(filename)\n",
    "        elif search_word == \"STOP\":\n",
    "            break\n",
    "        else:\n",
    "            print(\"I don't think we have what you typed, let's try again.\")\n",
    "            pass\n",
    "\n",
    "\n",
    "\n",
    "def find_DNA(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, dna_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"//\"):        # red flag\n",
    "                dna_flag = False\n",
    "            if dna_flag:\n",
    "                result += ''.join(line.split()[1:])\n",
    "            if line.startswith(\"ORIGIN\"):    # green flag\n",
    "                dna_flag = True\n",
    "        if dna_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(f\"\\nThe DNA sequence of the translated gene is:\\n{result}\\n\")\n",
    "\n",
    "def find_translated(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, tran_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            # this is the red line\n",
    "            # put the red line at the top so it is not included in the result :)\n",
    "            # this way the results stop at the correct place.\n",
    "            if line.startswith(\"exon\"):\n",
    "                tran_flag = False\n",
    "            if tran_flag:\n",
    "                # carful here, the keyword in the file is translation, not translated :).\n",
    "                result += line.strip(\"/translation=\").strip('\"').strip()\n",
    "            # this is the green line\n",
    "            # carful here with, remember the correct word is translation.\n",
    "            if line.startswith(\"/translation=\"):\n",
    "                tran_flag = True\n",
    "        if tran_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(f\"\\nThe AA sequence of the translated gene is:\\n{result}\\n\")\n",
    "\n",
    "\n",
    "def find_medline(filename):\n",
    "    medlines = []\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"MEDLINE\"):\n",
    "                medlines.append(line.strip(\"MEDLINE\").strip())\n",
    "            else:\n",
    "                pass\n",
    "    if len(medlines) > 1:\n",
    "        are = \"are\"\n",
    "    else:\n",
    "        are = \"is\"\n",
    "    print(f\"The medlines {are}: {medlines}\")\n",
    "\n",
    "\n",
    "def find_an_df_og(filename, search_word):\n",
    "    with open (f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # line.strip() will help with key word search like ORGANISM line, which starts with two spaces.\n",
    "            line = line.strip()\n",
    "            # the find start parts can be made into a smaller programme, but not that fit for what ex1 asks for.\n",
    "            # the way to do this is to ask the user for a string, that equals to \"ACCESSION\", \"DEFINITION\", ...etc\n",
    "            # search_word.split()[0] is useful for cases when we write accession number.\n",
    "            if line.startswith(f\"{search_word.split()[0].upper()}\"):\n",
    "                result = line.strip(f\"{search_word.split()[0].upper()}\").strip()\n",
    "                print(f\"The {search_word} is {result}.\")\n",
    "            elif len(search_word) == 0:\n",
    "                print(\"Empty search term!\")\n",
    "            else:\n",
    "                # using pass here for now\n",
    "                pass\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "804ce34b",
   "metadata": {},
   "source": [
    "5\\. Improve exercise 5.8 using all you have learned. The program shall now take a DNA FASTA file (asking interactively for it), and reverse and complement all entries in the file. There can be more than one entry, study dna7.fsa. Hint: Use substitution or translation for complementing the DNA. I will point out that the reading and writing of fasta files with many entries is a regular occurrence in bioinformatics (and exam), so be sure to get it right. Many people mistakenly believe that they should use a form of stateful parsing with a flag for this - doing so confuses the issue, so abstain from that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9720eb22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a DNA FASTA file name, such as dna.fsa, dna7.fsa: dna7.fsa\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "Summary: Keep the first line and reverse complement the sequence\n",
    "\n",
    "can utilise translate string for complementing the DNA.\n",
    "make a main function first\n",
    "    prompts user for DNA FASTA file name and saves that into a variable.\n",
    "    first with open an outfile, where we will write the reverse_comp into\n",
    "        second with open an infile, i.e. the file specified by the user, from which we read the information\n",
    "            for line in infile\n",
    "                strip the lines of unneeded white space\n",
    "                if the line starts with \">\":\n",
    "                    we write it to the outfile directly.\n",
    "                else:\n",
    "                    we run the reverse_complement function on the line. The output of this function is written into the outfile.\n",
    "\n",
    "what the reverse_complement function does:\n",
    "    takes in one_line\n",
    "    creates a translationTable for DNA nucleotide complementarity.\n",
    "    uses the .translate() method to translate the string in the given line.\n",
    "    converts this string into a list\n",
    "    uses .rever() to reverse this list inplace\n",
    "    uses join to save this list as a string variable, rev_comp_one_line\n",
    "    this small function returns the string variable, rev_comp_one_line\n",
    "\n",
    "The output file has each of the \"headerlines\" that begin with \">\" intact,\n",
    "followed by DNA nucleotides presented in the same format as before, only that now they are the reverse complements of the original :)\n",
    "\"\"\"\n",
    "# code\n",
    "def main():\n",
    "    usr_file = input(\"Please enter a DNA FASTA file name, such as dna.fsa, dna7.fsa: \")\n",
    "    with open(f\"{usr_file}\".split(\".\")[0]+\"_reverse_comp.fsa\", \"w\") as outfile:\n",
    "        with open(f\"{usr_file}\", \"r\") as infile:\n",
    "            for usr_line in infile:\n",
    "                usr_line = usr_line.strip()\n",
    "                if usr_line.startswith(\">\"):\n",
    "                    outfile.write(usr_line)\n",
    "                    outfile.write(\"\\n\")\n",
    "                else:\n",
    "                    outfile.write(rev_complement(usr_line))\n",
    "                    outfile.write(\"\\n\")\n",
    "\n",
    "def rev_complement(one_line):\n",
    "    translationTable = str.maketrans(\"ATCG\", \"TAGC\")\n",
    "    complement_one_line = one_line.translate(translationTable)\n",
    "    comp_list = list(complement_one_line)\n",
    "    comp_list.reverse()\n",
    "    rev_comp_one_line = ''.join(comp_list[:])\n",
    "    return rev_comp_one_line\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa57e45",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
