{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ab39ba30",
   "metadata": {},
   "source": [
    "## L7: Stateful Parsing\n",
    "Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5001f451",
   "metadata": {},
   "source": [
    "The following 5 exercises deal with SwissProt. The file sprot1.dat is a SwissProt database entry. Study it carefully. Locate the SwissProt ID (SP96_DICDI), the accession number (P14328) and the amino acid sequence (MRVLLVLVAC....TTTATTTATS). There are other entries ( sprot2.dat, sprot3.dat, sprot4.dat). Your programs should work on those, too. Also your programs must solve all the problems in ONE reading of the file. It is acceptable if you just hand in one program that solves 1 to 4. 5 is separate. To be clear - this exercise is only about studying and understanding the file format.\n",
    "\n",
    "1\\. Make a program that reads the ID and prints it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f464448",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter sprot file name: sprot3.dat\n",
      "ACH4_CHICK\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "\n",
    "def main():\n",
    "    file_to_work_on = input(\"Enter sprot file name: \")\n",
    "    read_print_ID(file_to_work_on)\n",
    "\n",
    "def read_print_ID(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "\n",
    "    (swissp_id, idflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"ID\":\n",
    "            idflag = False\n",
    "        if line[:2] == \"ID\": # this is the \"Green\", reposition later.\n",
    "            idflag = True\n",
    "        if idflag == True:\n",
    "            swissp_id += line[2:18].strip() # this slice should be big enough\n",
    "\n",
    "\n",
    "    if idflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_id)\n",
    "    \n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebe94697",
   "metadata": {},
   "source": [
    "2\\. Add the following functionality to the program: Read the accession number and print it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "319c407a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter sprot file name: sprot3.dat\n",
      "ACH4_CHICK\n",
      "P09482\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "\n",
    "def main():\n",
    "    file_to_work_on = input(\"Enter sprot file name: \")\n",
    "    read_print_ID(file_to_work_on)\n",
    "    read_print_AC(file_to_work_on)\n",
    "\n",
    "def read_print_ID(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "\n",
    "    (swissp_id, idflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"ID\":\n",
    "            idflag = False\n",
    "        if line[:2] == \"ID\": # this is the \"Green\", reposition later.\n",
    "            idflag = True\n",
    "        if idflag == True:\n",
    "            swissp_id += line[2:18].strip() # this slice should be big enough\n",
    "\n",
    "\n",
    "    if idflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_id)\n",
    "\n",
    "\n",
    "def read_print_AC(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ac, acflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"AC\":\n",
    "            acflag = False\n",
    "        if line[:2] == \"AC\":\n",
    "            acflag = True\n",
    "        if acflag == True:\n",
    "            for item in line[2:].strip().strip(\";\").split(\";\"):\n",
    "                swissp_ac += item    # sometimes there are multiple accession numbers\n",
    "\n",
    "    if acflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, ca not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_ac)\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "634c5239",
   "metadata": {},
   "source": [
    "3\\. Add the following functionality to the program: Read the amino acid sequence and print it. You really should use Stateful Parsing in this exercise. Maybe check the video.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "af4da9ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter sprot file name: sprot2.dat\n",
      "ASM_HUMAN\n",
      "P17405 P17406 Q13811 Q16837 Q16841\n",
      "MPRYGASLRQ SCPRSGREQG QDGTAGAPGL LWMGLVLALA LALALALSDS RVLWAPAEAHPLSPQGHPAR LHRIVPRLRD VFGWGNLTCP ICKGLFTAIN LGLKKEPNVA RVGSVAIKLCNLLKIAPPAV CQSIVHLFED DMVEVWRRSV LSPSEACGLL LGSTCGHWDI FSSWNISLPTVPKPPPKPPS PPAPGAPVSR ILFLTDLHWD HDYLEGTDPD CADPLCCRRG SGLPPASRPGAGYWGEYSKC DLPLRTLESL LSGLGPAGPF DMVYWTGDIP AHDVWHQTRQ DQLRALTTVTALVRKFLGPV PVYPAVGNHE STPVNSFPPP FIEGNHSSRW LYEAMAKAWE PWLPAEALRTLRIGGFYALS PYPGLRLISL NMNFCSRENF WLLINSTDPA GQLQWLVGEL QAAEDRGDKVHIIGHIPPGH CLKSWSWNYY RIVARYENTL AAQFFGHTHV DEFEVFYDEE TLSRPLAVAFLAPSATTYIG LNPGYRVYQI DGNYSRSSHV VLDHETYILN LTQANIPGAI PHWQLLYRARETYGLPNTLP TAWHNLVYRM RGDMQLFQTF WFLYHKGHPP SEPCGTPCRL ATLCAQLSARADSPALCRHL MPDGSLPEAQ SLWPRPLFC\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "\n",
    "def main():\n",
    "    file_to_work_on = input(\"Enter sprot file name: \")\n",
    "    read_print_ID(file_to_work_on)\n",
    "    read_print_AC(file_to_work_on)\n",
    "    read_print_AA(file_to_work_on)\n",
    "\n",
    "def read_print_ID(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "\n",
    "    (swissp_id, idflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"ID\":\n",
    "            idflag = False\n",
    "        if line[:2] == \"ID\": # this is the \"Green\", reposition later.\n",
    "            idflag = True\n",
    "        if idflag == True:\n",
    "            swissp_id += line[2:18].strip() # this slice should be big enough\n",
    "\n",
    "\n",
    "    if idflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_id)\n",
    "\n",
    "\n",
    "def read_print_AC(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ac, acflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"AC\":\n",
    "            acflag = False\n",
    "        if line[:2] == \"AC\":\n",
    "            acflag = True\n",
    "        if acflag == True:\n",
    "            for item in line[2:].strip().strip(\";\").split(\";\"):\n",
    "                swissp_ac += item    # sometimes there are multiple accession numbers\n",
    "\n",
    "    if acflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, ca not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_ac)\n",
    "\n",
    "\n",
    "def read_print_AA(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    # The structure for this code is the most similar to that of the lecture example.\n",
    "    (swissp_aa, aaflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:-1] == \"//\":    # This is the \"Red\" line.\n",
    "            aaflag = False\n",
    "        if aaflag == True:\n",
    "            swissp_aa += line[:-1].strip()\n",
    "        if line[:2] == \"SQ\":    # This is the \"Green\" line\n",
    "            aaflag = True\n",
    "\n",
    "    if aaflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_aa)\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b57bf54",
   "metadata": {},
   "source": [
    "4\\. Add the following functionality to the program: Verification of amino acid number. This means extract the number from the SQ line (example: SQ SEQUENCE 629 AA;) and check that the amino acid sequence has that number of residues. It should be the program that determines if something is wrong - not the user. Imagine that before you go home, you set the computer to run through a million swisprot entries. The next day, you must be able to see what failed. In a sense you don't care about what succeeded, as that is the common case. You care about what failed, because it is here you must take action."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1f400945",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter sprot file name: sprot1.dat\n",
      "SP96_DICDI\n",
      "P14328\n",
      "MRVLLVLVACLTYFSGGALAQSCSSYSGDNCPSTCFQGSYDIPCGAQVKYCTEMKDNCGEGGDVKCWKDGSNLPVQTWSSCVPSELFGPNGKFKPSEIPNSSNCPTNCENGVEWVNLCGLSCDAKTACCPDVCQCKGGQTSGGSTTGSQTSGGSTSGGSTTGSQTSGGSTTGSQTSGSQTSAGSCSNTQCPNGFYCQVQGNNAVCVPQQSSTSGGHQNDPCDTVQCPYGYSCESRDGFEAKCTRDEDEPTHRPTHRPKPPHDSDKYLCDNVHCPRGYKCNAKNGVAKCIAGYEIPRVCRNIQCPTGYRCEDHNRNPICVLEERENPDNCLTCNDVNCEASGLVCVMTRARCKVGAAKCCDVQPTCIKPSTIAGSTIASIASTIASTGSTGATSPCSVAQCPTGYVCVAQNNVAVSLPRPTTTTGSTSDSSALGSTSESSASGSSAVSSSASGSSAASSSPSSSAASSSPSSSAASSSPSSSAASSSPSSSASSSSSPSSSASSSSAPSSSASSSSAPSSSASSSSASSSSASSAATTAATTIATTAATTTATTTATTATTTATTTATTTAATIATTTAATTTATTTATTATTTATTTATS\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "\n",
    "def main():\n",
    "    file_to_work_on = input(\"Enter sprot file name: \")\n",
    "    read_print_ID(file_to_work_on)\n",
    "    read_print_AC(file_to_work_on)\n",
    "    print(f\"{read_AA(file_to_work_on)}\")\n",
    "    program_AA_count = (len(f\"{read_AA(file_to_work_on)}\"))\n",
    "    # verify AA counts are correct\n",
    "    if AA_count(file_to_work_on) != program_AA_count:\n",
    "        print(\"Programme's AA counts does not equal SQ line AA count, please check what happened!\")\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "def read_print_ID(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "\n",
    "    (swissp_id, idflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"ID\":\n",
    "            idflag = False\n",
    "        if line[:2] == \"ID\": # this is the \"Green\", reposition later.\n",
    "            idflag = True\n",
    "        if idflag == True:\n",
    "            swissp_id += line[2:18].strip() # this slice should be big enough\n",
    "\n",
    "\n",
    "    if idflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_id)\n",
    "\n",
    "\n",
    "def read_print_AC(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ac, acflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"AC\":\n",
    "            acflag = False\n",
    "        if line[:2] == \"AC\":\n",
    "            acflag = True\n",
    "        if acflag == True:\n",
    "            for item in line[2:].strip().strip(\";\").split(\";\"):\n",
    "                swissp_ac += item    # sometimes there are multiple accession numbers\n",
    "\n",
    "    if acflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, ca not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    print(swissp_ac)\n",
    "\n",
    "\n",
    "def read_AA(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    # The structure for this code is the most similar to that of the lecture example.\n",
    "    (swissp_aa, aaflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:-1] == \"//\":    # This is the \"Red\" line.\n",
    "            aaflag = False\n",
    "        if aaflag == True:\n",
    "            swissp_aa += line[:-1].replace(\" \", \"\").replace(\"\\n\", \"\").strip()\n",
    "        if line[:2] == \"SQ\":    # This is the \"Green\" line.\n",
    "            aaflag = True\n",
    "\n",
    "    if aaflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "\n",
    "    infile.close()\n",
    "    return(swissp_aa)\n",
    "\n",
    "\n",
    "def AA_count(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_aan, aanflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"SQ\": # This is the \"Red\" line.\n",
    "            aanflag = False\n",
    "        if line[:2] == \"SQ\": # This is the \"Green\" line.\n",
    "            aanflag = True\n",
    "        if aanflag == True:\n",
    "            swissp_aan += line.split()[2]\n",
    "            swissp_aan = int(swissp_aan)\n",
    "\n",
    "    if aanflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "\n",
    "    infile.close()\n",
    "    return(swissp_aan)\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0132212",
   "metadata": {},
   "source": [
    "5\\. Now that you have the ID, accession number and AA sequence save it to a file sprot.fsa in FASTA format. Look in the file dna.fsa for an example of FASTA. Notice the first line starts with > and immediately after comes an unique identifier, like an accession number or a SwissProt ID. Any other data must be on the header line only, but in free format. Sequence data is on the following lines.\n",
    "Notice that this exercise incorporates the previous 4, but uses the result in a slightly different way. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dbcfac4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter sprot*.dat filename, for example, sprot2.dat: sprot1.dat\n"
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    # write the result of these functions to a file of the same name, with but with suffix fsa instead.\n",
    "    file_to_work_on = input(\"Enter sprot*.dat filename, for example, sprot2.dat: \")\n",
    "    ID = \">\" + extract_ID(file_to_work_on)\n",
    "    AAN = AA_count(file_to_work_on)\n",
    "    OS = extract_OS(file_to_work_on)\n",
    "    GN = extract_GN(file_to_work_on)\n",
    "    AA = read_AA(file_to_work_on)\n",
    "    header = [ID, AAN, OS, GN]\n",
    "    \n",
    "    \"\"\"\n",
    "    Example output with sprot2.dat:\n",
    "    >ASM_HUMAN 629 AA Homo sapiens gene=\"SMPD1 OR ASM\"\n",
    "    amino acid sequence, with out spaces, new line after 60 AAs.\n",
    "    \"\"\"\n",
    "\n",
    "    outfile_name = file_to_work_on.replace(\"dat\", \"fsa\")\n",
    "    with open (f\"{outfile_name}\", \"w\") as outfile:\n",
    "        for item in header:\n",
    "            outfile.write(f\"{item} \") # adding the space here as the separater :)\n",
    "        outfile.write(f\"{AA}\")   # when you use write again, we get a second line automatically.\n",
    "\n",
    "# extracts the .dat file's swissp_ID\n",
    "def extract_ID(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "\n",
    "    (swissp_id, idflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"ID\":\n",
    "            idflag = False\n",
    "        if line[:2] == \"ID\": # this is the \"Green\", reposition later.\n",
    "            idflag = True\n",
    "        if idflag == True:\n",
    "            swissp_id += line[2:18].strip() # this slice should be big enough\n",
    "\n",
    "\n",
    "    if idflag:    # if it is still set to True\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    return swissp_id\n",
    "\n",
    "# extracts the .dat file's amino acid count\n",
    "def AA_count(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_aan, aanflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"SQ\": # This is the \"Red\" line.\n",
    "            aanflag = False\n",
    "        if line[:2] == \"SQ\": # This is the \"Green\" line.\n",
    "            aanflag = True\n",
    "        if aanflag == True:\n",
    "            swissp_aan += line.split()[2]\n",
    "            swissp_aan = swissp_aan + \" AA\"\n",
    "\n",
    "    if aanflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "\n",
    "    infile.close()\n",
    "    return swissp_aan\n",
    "\n",
    "# extracts the .dat file's swissp_OS, the species name only.\n",
    "def extract_OS(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "\n",
    "    (swissp_os, osflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"OS\":\n",
    "            osflag = False\n",
    "        if line[:2] == \"OS\":\n",
    "            osflag = True\n",
    "        if osflag == True:\n",
    "            line = line[2:].strip().strip(\".\").split()\n",
    "            for word in line:\n",
    "                if word.isalnum():\n",
    "                    swissp_os += word + \" \"\n",
    "                else:\n",
    "                    pass\n",
    "\n",
    "    if osflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "    infile.close()\n",
    "    return swissp_os.strip()    #.strip() removes the extra space at the end\n",
    "\n",
    "# extracts the .dat file's gene name\n",
    "def extract_GN(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_gn, gnflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"GN\":\n",
    "            gnflag = False\n",
    "        if line[:2] == \"GN\":\n",
    "            gnflag = True\n",
    "        if gnflag == True:\n",
    "            swissp_gn += line[2:].strip().strip(\".\")\n",
    "            swissp_gn = 'gene='+ '\"'f\"{swissp_gn}\"'\"'\n",
    "\n",
    "    infile.close()\n",
    "    return swissp_gn\n",
    "\n",
    "# extracts the .dat file's amino acid sequence\n",
    "def read_AA(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    # The structure for this code is the most similar to that of the lecture example.\n",
    "    (swissp_aa, aaflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:-1] == \"//\":    # This is the \"Red\" line.\n",
    "            aaflag = False\n",
    "        if aaflag == True:\n",
    "            swissp_aa += line[:-1].replace(\" \", \"\").replace(\"\\n\", \"\").strip()\n",
    "        if line[:2] == \"SQ\":    # This is the \"Green\" line.\n",
    "            aaflag = True\n",
    "\n",
    "    if aaflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "\n",
    "\n",
    "    infile.close()\n",
    "    swissp_aa_formated = \"\"\n",
    "    char_count = 0\n",
    "    for char in swissp_aa:\n",
    "        if char_count % 61 == 0:\n",
    "            swissp_aa_formated += \"\\n\"\n",
    "            char_count += 1\n",
    "        else:\n",
    "            swissp_aa_formated += char\n",
    "            char_count +=1\n",
    "\n",
    "    return swissp_aa_formated\n",
    "\n",
    "\n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
