{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "837b6405",
   "metadata": {},
   "source": [
    "## L11: Sets\n",
    "Exercises for Extra Practice\n",
    "\n",
    "A k-mer is a piece of sequence k units long. It can be both protein and DNA sequence. As an example I have this DNA sequence: GACTAC. It contains the following 4-mers: GACT, ACTA, CTAC."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d2e0ca",
   "metadata": {},
   "source": [
    "1\\. From a genbank file like data1-4.gb extract the DNA sequence (been there - done that). Now insert all 5-mers in the DNA sequence into a set. Compute the following numbers: how many did you insert in the set, how many are in the set, how many different 5-mers could there possibly be. Display. It seems like the two first numbers are the same, but this is not guaranteed to be true, why?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6ebf128a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Which gene-bank file do you want to extract the DNA sequence from?\n",
      "Choose one from data1-4.gb (STOP to quit): data1.gb\n",
      "We will give you a lot of n-mers, specify your n: 5\n",
      "There were 4040 5-mer insertions to the set.\n",
      "The set contains 764 5-mers in total.\n",
      "There are 1024 possible unique 5-mers in total.\n",
      "Out of all possible unique 5-mers, the 5-mers from data1.gb included all but the following: \n",
      "['aaaac', 'aaaat', 'aaacg', 'aaata', 'aaatc', 'aacag', 'aacgg', 'aacgt', 'aactg', 'aactt', 'aatac', 'aatag', 'aatat', 'aatca', 'aatcc', 'aatcg', 'aatta', 'aattc', 'acaag', 'acagt', 'accgt', 'acgat', 'acgcg', 'acggg', 'acgta', 'acgtt', 'actaa', 'actta', 'agaaa', 'agcga', 'agcta', 'agtcg', 'agttg', 'agttt', 'ataac', 'ataag', 'ataat', 'atacc', 'atacg', 'ataga', 'atata', 'atatc', 'atatg', 'atatt', 'atcat', 'atccc', 'atcga', 'atcgc', 'atcgg', 'atcgt', 'atcta', 'atgaa', 'atgat', 'atgta', 'atgtt', 'attaa', 'attac', 'attag', 'attat', 'attcg', 'attct', 'attga', 'attgg', 'attta', 'caaaa', 'caata', 'caatc', 'cacaa', 'cacat', 'cagtt', 'cataa', 'catac', 'catat', 'catcg', 'catga', 'catta', 'ccaat', 'cccgt', 'ccgaa', 'ccgat', 'ccgcg', 'ccggt', 'ccgta', 'ccgtg', 'ccgtt', 'cctat', 'cctta', 'cgaaa', 'cgaat', 'cgacg', 'cgact', 'cgagc', 'cgata', 'cgatc', 'cgatg', 'cgatt', 'cgcac', 'cgcga', 'cgcgc', 'cgcgt', 'cgcta', 'cgctt', 'cggac', 'cggat', 'cggcg', 'cgggt', 'cggta', 'cggtc', 'cggtt', 'cgtaa', 'cgtac', 'cgtag', 'cgtat', 'cgtcg', 'cgtgc', 'cgtgt', 'cgtta', 'cgttc', 'cgttt', 'ctacg', 'ctatc', 'ctccg', 'cttaa', 'cttac', 'cttag', 'cttat', 'cttta', 'ctttc', 'gaaaa', 'gaaac', 'gaatt', 'gagta', 'gataa', 'gatat', 'gatcg', 'gatgt', 'gatta', 'gattg', 'gattt', 'gcaag', 'gcata', 'gcgat', 'gcggt', 'gcgta', 'gctac', 'gctag', 'gctat', 'gctta', 'ggatt', 'ggcta', 'gggta', 'ggtac', 'ggtat', 'ggtcg', 'ggttt', 'gtaac', 'gtaat', 'gtaca', 'gtacg', 'gtaga', 'gtagt', 'gtata', 'gtatc', 'gtatt', 'gtcaa', 'gtcga', 'gtcgg', 'gtcgt', 'gttac', 'gttcg', 'gttta', 'gtttc', 'gtttg', 'taaac', 'taaat', 'taaca', 'taacg', 'taact', 'taagc', 'taata', 'taatc', 'taatt', 'tacaa', 'tacag', 'taccg', 'tacga', 'tacgc', 'tacgg', 'tacgt', 'tacta', 'tactt', 'tagaa', 'tagat', 'tagca', 'tagcg', 'tagga', 'tagta', 'tagtt', 'tatac', 'tatag', 'tatat', 'tatca', 'tatcc', 'tatcg', 'tatct', 'tatga', 'tatgc', 'tatgt', 'tatta', 'tattc', 'tattg', 'tcaac', 'tcaat', 'tcata', 'tccgc', 'tccgt', 'tcgaa', 'tcgac', 'tcgat', 'tcgca', 'tcgcg', 'tcgct', 'tcggt', 'tcgta', 'tcgtc', 'tcgtt', 'tctag', 'tctcg', 'tctta', 'tcttt', 'tgaaa', 'tgaat', 'tgata', 'tgatt', 'tggta', 'tgtaa', 'tgtat', 'ttaac', 'ttaat', 'ttaca', 'ttacc', 'ttacg', 'ttact', 'ttagc', 'ttagg', 'ttata', 'ttatc', 'ttatg', 'ttcac', 'ttcat', 'ttcga', 'ttcgc', 'ttcgt', 'ttgag', 'ttgta', 'tttac', 'tttat', 'tttca', 'tttcg', 'tttgg']\n",
      "Which gene-bank file do you want to extract the DNA sequence from?\n",
      "Choose one from data1-4.gb (STOP to quit): data2.gb\n",
      "We will give you a lot of n-mers, specify your n: 5\n",
      "There were 38678 5-mer insertions to the set.\n",
      "The set contains 1020 5-mers in total.\n",
      "There are 1024 possible unique 5-mers in total.\n",
      "Out of all possible unique 5-mers, the 5-mers from data2.gb included all but the following: \n",
      "['cgcgt', 'cgtcg', 'gacgc', 'gcgta']\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "Interrupted by user",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/9c/bz3_nczs67l3rlf5shp54rjh0000gn/T/ipykernel_53916/3965397329.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     94\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     95\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 96\u001b[0;31m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/9c/bz3_nczs67l3rlf5shp54rjh0000gn/T/ipykernel_53916/3965397329.py\u001b[0m in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m         \u001b[0musr_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Which gene-bank file do you want to extract the DNA sequence from?\\nChoose one from data1-4.gb (STOP to quit): \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0musr_file\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"STOP\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m             \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/anaconda3/lib/python3.9/site-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36mraw_input\u001b[0;34m(self, prompt)\u001b[0m\n\u001b[1;32m   1004\u001b[0m                 \u001b[0;34m\"raw_input was called, but this frontend does not support input requests.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1005\u001b[0m             )\n\u001b[0;32m-> 1006\u001b[0;31m         return self._input_request(\n\u001b[0m\u001b[1;32m   1007\u001b[0m             \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprompt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1008\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_ident\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"shell\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/anaconda3/lib/python3.9/site-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36m_input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m   1049\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1050\u001b[0m                 \u001b[0;31m# re-raise KeyboardInterrupt, to truncate traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1051\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Interrupted by user\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1052\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1053\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwarning\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid Message:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_info\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: Interrupted by user"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "prompt user for a file within data1-4.gb\n",
    "with open the file\n",
    "    use stateful parsing to extract the DNA sequence into one string..\n",
    "\n",
    "trying to write a function for 4-mers and expand it to apply to 5-mer\n",
    "    will make use of slicing and a for loop.\n",
    "\n",
    "number of possibl different 5-mers can be calculated by editing my fuction to calculate number of possible codons.\n",
    "now instead of making a string of len = 3 for each codon using different combinations of \"ATCG\",\n",
    "I will make a strng of len = 5 for each codon using different combinations of \"ATCG\".\n",
    "\n",
    "\"\"\"\n",
    "def main():\n",
    "    while True:\n",
    "        usr_file = input(\"Which gene-bank file do you want to extract the DNA sequence from?\\nChoose one from data1-4.gb (STOP to quit): \")\n",
    "        if usr_file == \"STOP\":\n",
    "            break\n",
    "        else:\n",
    "            usr_seq = get_dna_seq(usr_file)\n",
    "            usr_n = int(input(\"We will give you a lot of n-mers, specify your n: \"))\n",
    "            (n_insertions, usr_nmer_set) = n_mer_set(usr_seq, usr_n)\n",
    "            print(f\"There were {n_insertions} 5-mer insertions to the set.\")\n",
    "            print(f\"The set contains {len(usr_nmer_set)} 5-mers in total.\")\n",
    "\n",
    "            # how many different n-mers can there possibly be?\n",
    "            # modify the codon generator to make all uniqe 5-mers\n",
    "            # includes 5-mers where all 5 nucleotides are the same, i.e. \"AAAAA\"\n",
    "            all_5mers_set = make_all_5mers()\n",
    "            print(f\"There are {len(all_5mers_set)} possible unique 5-mers in total.\")\n",
    "            # check out which unique 5mers were not part of the n_mer_set from a given file:\n",
    "            print(f\"Out of all possible unique 5-mers, the 5-mers from {usr_file} included all but the following: \")\n",
    "            print(sorted(all_5mers_set.difference(usr_nmer_set)))\n",
    "\n",
    "\n",
    "# given seq string and n_mer number\n",
    "# returns possible n_mers\n",
    "def get_dna_seq(file):\n",
    "    with open(f\"{file}\", \"r\") as infile:\n",
    "        (dna_seq, dna_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            if line[:-1] == \"//\":\n",
    "                dna_flag = False\n",
    "            if dna_flag:\n",
    "                dna_seq += \"\".join(line.strip().split()[1:])\n",
    "            if \"ORIGIN\" in line:\n",
    "                dna_flag = True\n",
    "    if dna_flag:\n",
    "        raise ValueError(\"Format is not right. Cannot trust result.\")\n",
    "    if len(dna_seq) == 0:\n",
    "        raise ValueError(\"No data found. Cannot trust result.\")\n",
    "    # return dna_seq instead of print dna_seq here :)\n",
    "    # so that in can be used as input for the n_mer function in main.\n",
    "    return(dna_seq)\n",
    "\n",
    "\n",
    "# This one addes each n_mer into a set, and counts the number of times of insertions.\n",
    "# It does not print out each n_mer as they are made.\n",
    "def n_mer_set(seq, n):\n",
    "    new_set = set()\n",
    "    # counts \"how many n-mers\" I inserted into the set\n",
    "    insertion_counts = 0\n",
    "    # this is to prevent len(seq) from being interpreted as a string of len = 3. \n",
    "    len_seq = len(f\"{seq}\")\n",
    "\n",
    "    # a list of all the possible starting positions for the n-mer\n",
    "    # +1 to adjust for the index, try with a toy example if you forget:\n",
    "    start_positions = list(range(0,(len_seq - n) + 1))    \n",
    "    for start_position in start_positions:    # this is the starting value\n",
    "        end = start_position + n  # notice this is the correct index format for slicing.\n",
    "        new_set.add(seq[start_position:end])\n",
    "        insertion_counts += 1\n",
    "    \n",
    "    return(insertion_counts, new_set)\n",
    "\n",
    "\n",
    "# code adopted from https://www.geeksforgeeks.org/python-program-to-print-all-possible-combinations-from-the-three-digits/\n",
    "# to make all 5-mers:\n",
    "def make_all_5mers():\n",
    "    L = \"atcg\"    # remember to set this to lower case\n",
    "    fivemers_set = set()\n",
    "    len_nucleotides = len(L)\n",
    "    for i in range(len_nucleotides):\n",
    "        for j in range(len_nucleotides):\n",
    "            for k in range(len_nucleotides):\n",
    "                for l in range(len_nucleotides):\n",
    "                    for m in range(len_nucleotides):\n",
    "                        fivemer = L[i]+L[j]+L[k]+L[l]+L[m]\n",
    "                        fivemers_set.add(fivemer)\n",
    "    # sanity check\n",
    "    return(fivemers_set)\n",
    "    \n",
    "    \n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6290f0d6",
   "metadata": {},
   "source": [
    "3\\. Calculate the overlap of 5-mers between any two of the data1-4.gb files. Just ask for two filenames and calculate for these."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "51bde8e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter two file names from data1-4.gb, separated by space (STOP NOW to exit): data1.gb data2.gb\n",
      "Please enter the value of n for your n-mer: 5\n",
      "data1.gb and data2.gb contain 763 5-mer overlaps\n",
      "Please enter two file names from data1-4.gb, separated by space (STOP NOW to exit): data2.gb data3.gb\n",
      "Please enter the value of n for your n-mer: 5\n",
      "data2.gb and data3.gb contain 675 5-mer overlaps\n",
      "Please enter two file names from data1-4.gb, separated by space (STOP NOW to exit): STOP NOW\n"
     ]
    }
   ],
   "source": [
    "# pseudo code:\n",
    "\"\"\"\n",
    "prompt user for two file names\n",
    "\n",
    "extract the dna sequence from the file\n",
    "\n",
    "run each file's dna seq through a function that will return a set of 5-mers\n",
    "\n",
    "use the set method .intersection to find out the 5-mers that do overlap, and how many there are.\n",
    "\"\"\"\n",
    "\n",
    "def main():\n",
    "# prompt user for two filenames:\n",
    "    while True:\n",
    "        # use .spit() to save the two filename inputs as two separate variables\n",
    "        try:\n",
    "            usr_file1, usr_file2 = input(\"Please enter two file names from data1-4.gb, separated by space (STOP NOW to exit): \").split()\n",
    "        except ValueError as error:\n",
    "            print(\"Please enter two filenames.\")\n",
    "        except FileNotFoundError:\n",
    "            print(\"Please enter two filenames from data1-4.gb\")\n",
    "        else:\n",
    "            two_usr_files = [usr_file1, usr_file2]\n",
    "            if two_usr_files == \"STOP NOW\".split():\n",
    "                break\n",
    "    \n",
    "        # extract dna seq of both files\n",
    "        two_usr_file_seqs = []\n",
    "        for usr_file in two_usr_files:\n",
    "            # make sure get_file_seq will return a str.\n",
    "            two_usr_file_seqs.append(get_dna_seq(usr_file))\n",
    "        \n",
    "        # generate 5-mers of both files\n",
    "        usr_n = int(input(\"Please enter the value of n for your n-mer: \"))\n",
    "        two_usr_nmer_sets = []\n",
    "        for usr_seq in two_usr_file_seqs:\n",
    "            two_usr_nmer_sets.append(n_mer_set(usr_seq, usr_n))\n",
    "        \n",
    "        # sanity check to make sure we have sets within a list:\n",
    "        # print(type(two_usr_nmers_sets[0]))\n",
    "        n_mer_set_intersection = two_usr_nmer_sets[0].intersection(two_usr_nmer_sets[1])\n",
    "        print(f\"{usr_file1} and {usr_file2} contain {len(n_mer_set_intersection)} {usr_n}-mer overlaps\")\n",
    "    # before moving on to index the list above and find intersection.\n",
    "\n",
    "# modified from L11_ex1's n_mer_set.\n",
    "# takes DNA seq strings and n_mer's n as input\n",
    "# only returns a set of n_mers\n",
    "# does not return insertion coutns or lenth count.\n",
    "def n_mer_set(seq, n):\n",
    "    new_set = set()\n",
    "    # this variable prevents seq from being interpreted as a string of len 3.\n",
    "    len_seq = len(f\"{seq}\")\n",
    "\n",
    "    start_positions = list(range(0, (len_seq - n)+1))\n",
    "    for start_position in start_positions:\n",
    "        end = start_position + n\n",
    "        new_set.add(seq[start_position:end])\n",
    "    \n",
    "    return(new_set)\n",
    "\n",
    "\n",
    "# input genebank format filename\n",
    "# outputs one string of dna\n",
    "def get_dna_seq(file):\n",
    "    with open(f\"{file}\", \"r\") as infile:\n",
    "        (dna_seq, dna_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            if line[:-1] == \"//\":\n",
    "                dna_flag = False\n",
    "            if dna_flag:\n",
    "                dna_seq += \"\".join(line.strip().split()[1:])\n",
    "            if \"ORIGIN\" in line:\n",
    "                dna_flag = True\n",
    "    if dna_flag:\n",
    "        raise ValueError(\"Format is not right. Cannot trust result.\")\n",
    "    if len(dna_seq) == 0:\n",
    "        raise ValueError(\"No data found. Cannot trust result.\")\n",
    "    # return dna_seq instead of print dna_seq here.\n",
    "    # so that it can easily be used as an input for a different function.\n",
    "    return(dna_seq)\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "180245fe",
   "metadata": {},
   "source": [
    "3\\. Make a program that asks for a number (integer) then reads the dna7.fsa file (which contains insulin-like genes) and saves the entry selected by the number in selected.fsa and the rest of the entries into the file rest.fsa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cb5f20ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter the .fsa filename you want to look at (i.e. dna7.fsa): dna5.fsa\n",
      "Sorry, we only have dna7.fsa to work on, pleace try again and enter dna7.fsa\n",
      "Please enter the .fsa filename you want to look at (i.e. dna7.fsa): dna7.fsa\n",
      "Please enter the entry number you want to select from dna7.fsa: 7\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "main function:\n",
    "    prompt user for a number\n",
    "        convert it to an int.\n",
    "\n",
    "    prompt user for a file name (can only be dna7.fsa in this cse)\n",
    "    use fsa_entry_to_list function on file.\n",
    "    use list_to_files on fsa_entry_to_list and usr's selected number.\n",
    "        \n",
    "    function fsa_entry_to_list: input file name, outputs list of entries.\n",
    "        read the dna7.fsa file\n",
    "    put each entry into a list\n",
    "    using a list because lists are ordered\n",
    "\n",
    "    function list_to_files: input list of entries AND selected number, output two files.\n",
    "        access the correct index on the list to save the entry selected by the usr into selected.fsa,\n",
    "        can do this by returning the output of .pop(correct index)\n",
    "\n",
    "        save the list after doing .pop() into file rest.fsa\n",
    "\n",
    "return the main function\n",
    "\"\"\"\n",
    "\n",
    "def main():\n",
    "    # the filename will only be dna7.fsa in this case.\n",
    "    while True:\n",
    "        usr_filename = input(f\"Please enter the .fsa filename you want to look at (i.e. dna7.fsa): \")\n",
    "        if usr_filename == \"dna7.fsa\":\n",
    "            usr_number = int(input(f\"Please enter the entry number you want to select from {usr_filename}: \"))\n",
    "            usr_fsa_as_list = fsa_entry_to_list(usr_filename)\n",
    "            list_to_files(usr_fsa_as_list, usr_number)\n",
    "            # break here so we leave once we get wehat we need :)\n",
    "            break\n",
    "        else:\n",
    "            print(\"Sorry, we only have dna7.fsa to work on, pleace try again and enter dna7.fsa\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def list_to_files(entries_as_list, entry_number):\n",
    "    # make sure to use the correct index number:\n",
    "    index_number = entry_number - 1\n",
    "    # after pop, we have an list that was changed in-place.\n",
    "    selected_entry = entries_as_list.pop(index_number)\n",
    "    rest_entries_list = entries_as_list\n",
    "    with open(\"selected_entry.fsa\", \"w\") as selected_outfile:\n",
    "            # output file displayed in a pretty way because of correct when reading the file earlier.\n",
    "            # i.e. by not doing line = line.strip() when reading the file.\n",
    "            selected_outfile.write(selected_entry)\n",
    "    with open(\"rest_entries.fsa\", \"w\") as rest_outfile:\n",
    "        for entry in entries_as_list:\n",
    "            rest_outfile.write(entry) \n",
    "\n",
    "\n",
    "\n",
    "def fsa_entry_to_list(file):\n",
    "    seq = \"\"\n",
    "    with open(f\"{file}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # rember not to do line = line.strip() here\n",
    "            # it will help to keep your output file format pretty :)\n",
    "            seq += line\n",
    "    # thinking to use \">\" split, checking the output.\n",
    "    # use replace(\">\", \"@>\").split(\"@\") to retain > but also achieve correct splitting.\n",
    "    seq_list = seq.replace(\">\",\"@>\").split(\"@\")\n",
    "    # pop(0) removes the leading \"\" in the list, which we don't need.\n",
    "    seq_list.pop(0)\n",
    "    return(seq_list)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f443774",
   "metadata": {},
   "source": [
    "4\\. Now that you can select a fasta entry, then read the selected.fsa and create 5-mers from the sequence. Now read the entries from rest.fsa and for every entry create the 5-mers from the sequence. Report which sequence in rest.fsa had the greatest overlap (and how much overlap) with the selected sequence. This must be the sequence that looks most like the selected one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "76804ce0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter the .fsa filename you want to look at (i.e. dna7.fsa): dna7.fsa\n",
      "Please enter the entry number you want to select from dna7.fsa: 4\n",
      "Please enter the value of n for your n-mer: 5\n",
      "\n",
      "The following sequence has the greatest overlap with the selected sequence, with 52 overlapping 5-mers.\n",
      "{'GCAGA', 'TGGAT', 'GCGTG', 'GGTTG', 'GCACT', 'GTTGC', 'CACTG', 'GCTGC', 'GGAGG', 'TCAGT', 'AGTGC', 'GATCA', 'ACTGG', 'GGATC', 'TGCAC', 'CCCGG', 'GCACC', 'CAGAA', 'GAGGT', 'CCGGC', 'GAAGC', 'CACCA', 'TCTGC', 'CTGCA', 'ATCAG', 'AGCGT', 'CTGGA', 'ATTGT', 'CCAGC', 'TGTGG', 'CAGTG', 'GCATT', 'AGGTT', 'CATTG', 'TTGTG', 'GTGCT', 'CAGCA', 'CGGCA', 'GGCAG', 'AAGCG', 'ATCTG', 'TTGCC', 'ACCAG', 'CATCT', 'TGCCC', 'TGGAG', 'TGGCA', 'GCATC', 'TGCTG', 'CGTGG', 'GTGGC', 'AGCAT', 'GGCAT', 'AGAAG', 'GTGGA', 'GCCCG'}\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "you have the selected.fsa and rest.fsa in the directory already, they are outputs from the previous question.\n",
    "but the user might want to select a different line from what I selected while writting the previous exercise.\n",
    "So, the code includes that of the previous exercise, and contains an additinoal section,\n",
    "which implements the n_mer_set function I made in L11_ex1_extra.py,\n",
    "and uses .difference() to find the greatest overlap.\n",
    "\n",
    "read the selected file, extract the sequence.\n",
    "seq and follow the L11_ex2.extra.py or L11_ex1.extra.py file to create 5-mers from the sequence.\n",
    "\n",
    "read the rest file seq, extract the sequence into a list.\n",
    " and do the same, create the 5-mers from the sequences, put them in a list.\n",
    "\n",
    "initiate an empty string that will always be the intersection with the largest length, \n",
    "for each rest_sequence file in list of rest_sequences:\n",
    "    if len((rest_sequence).intersection(selected_sequence))) is greater than len(empty_sring).\n",
    "        then update empty string's content to that of the rest_sequence.\n",
    "\"\"\"\n",
    "# code\n",
    "def main():\n",
    "    # the filename will only be dna7.fsa in this case.\n",
    "    while True:\n",
    "        usr_filename = input(f\"Please enter the .fsa filename you want to look at (i.e. dna7.fsa): \")\n",
    "        if usr_filename == \"dna7.fsa\":\n",
    "            usr_number = int(input(f\"Please enter the entry number you want to select from {usr_filename}: \"))\n",
    "            usr_fsa_as_list = fsa_entry_to_list(usr_filename)\n",
    "            list_to_files(usr_fsa_as_list, usr_number)\n",
    "            break\n",
    "        else:\n",
    "            print(\"Sorry, we only have dna7.fsa to work on, pleace try again and enter dna7.fsa\")\n",
    "    # code here, outside the while loop will also operate.\n",
    "    # we know for sure the file names will be:\n",
    "    # selected_entry.fsa and rest_entries.fsa\n",
    "    # read the sequence into 5-mers for select.fsa:\n",
    "    selected_seq = \"\"\n",
    "    with open(\"selected_entry.fsa\", \"r\") as selected_infile:\n",
    "        for line in selected_infile:\n",
    "            if not line.startswith(\">\"):\n",
    "                selected_seq += line.strip()\n",
    "    # sanity check we have one entire string.\n",
    "    # print(selected_seq) worked :)\n",
    "    rest_entries_list = []\n",
    "    with open(\"rest_entries.fsa\", \"r\") as rest_infile:\n",
    "        for line in rest_infile:\n",
    "            if not line.startswith(\">\"):\n",
    "                rest_entries_list.append(line.strip())\n",
    "    # sanity check with have one list with seqs as items.\n",
    "    # print(rest_entries_list) worked :)\n",
    "\n",
    "\n",
    "    # make 5-mers for each using the seq_to_5-mer function\n",
    "    usr_n = int(input(\"Please enter the value of n for your n-mer: \"))\n",
    "    selected_n_mer_set = n_mer_set(selected_seq, usr_n)\n",
    "    \n",
    "    rest_n_mer_list = []\n",
    "    for rest_entry in rest_entries_list:\n",
    "        rest_n_mer_set = n_mer_set(rest_entry, usr_n)\n",
    "        # so now we have a list with many sets.\n",
    "        # and later we use a for loop to compare each set in this list.\n",
    "        rest_n_mer_list.append(rest_n_mer_set)\n",
    "\n",
    "    greatest_overlap_num = 0\n",
    "    for entry in rest_n_mer_list:\n",
    "        if len((selected_n_mer_set).intersection(entry)) > greatest_overlap_num:\n",
    "            greatest_overlap_num = len((selected_n_mer_set).intersection(entry))\n",
    "            greatest_overlap_entry = entry\n",
    "        else:\n",
    "            pass\n",
    "    \n",
    "    print(f\"\\nThe following sequence has the greatest overlap with the selected sequence, with {greatest_overlap_num} overlapping 5-mers.\")\n",
    "    print(greatest_overlap_entry)\n",
    "\n",
    "def n_mer_set(seq, n):\n",
    "    new_set = set()\n",
    "    # this variable prevents seq from being interpreted as a string of len 3.\n",
    "    len_seq = len(f\"{seq}\")\n",
    "\n",
    "    start_positions = list(range(0, (len_seq - n)+1))\n",
    "    for start_position in start_positions:\n",
    "        end = start_position + n\n",
    "        new_set.add(seq[start_position:end])\n",
    "    \n",
    "    return(new_set)\n",
    "\n",
    "\n",
    "\n",
    "def list_to_files(entries_as_list, entry_number):\n",
    "    # make sure to use the correct index number:\n",
    "    index_number = entry_number - 1\n",
    "    # after pop, we have an list that was changed in-place.\n",
    "    selected_entry = entries_as_list.pop(index_number)\n",
    "    rest_entries_list = entries_as_list\n",
    "    with open(\"selected_entry.fsa\", \"w\") as selected_outfile:\n",
    "            # output file displayed in a pretty way because of correct when reading the file earlier.\n",
    "            # i.e. by not doing line = line.strip() when reading the file.\n",
    "            selected_outfile.write(selected_entry)\n",
    "    with open(\"rest_entries.fsa\", \"w\") as rest_outfile:\n",
    "        for entry in entries_as_list:\n",
    "            rest_outfile.write(entry) \n",
    "\n",
    "\n",
    "\n",
    "def fsa_entry_to_list(file):\n",
    "    seq = \"\"\n",
    "    with open(f\"{file}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # rember not to do line = line.strip() here\n",
    "            # it will help to keep your output file format pretty :)\n",
    "            seq += line\n",
    "    # thinking to use \">\" split, checking the output.\n",
    "    # use replace(\">\", \"@>\").split(\"@\") to retain > but also achieve correct splitting.\n",
    "    seq_list = seq.replace(\">\",\"@>\").split(\"@\")\n",
    "    # pop(0) removes the leading \"\" in the list, which we don't need.\n",
    "    seq_list.pop(0)\n",
    "    return(seq_list)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
