{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f27b4cd2",
   "metadata": {},
   "source": [
    "## L10: Simple Pattern Matching\n",
    "Exercises for Extra Practice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "506fee2f",
   "metadata": {},
   "source": [
    "1\\. Given that you have a string containing some DNA (ask for input or hardcode it in the program), how would you check that it actually only contains ATCG using the methods of today's lesson?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7283280",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a DNA string that might contain non-ATCGs: ATGATGATXCGATCGATYCGATGCAZT\n",
      "The original DNA string contains letters other than ATCGs.\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "prompt user for DNA string.\n",
    "to check that it only contains ATCG using today's lesson (translate)\n",
    "we can:\n",
    "create a new empty string to capture non ATCG characters\n",
    "for each char in the original string\n",
    "    if it is in 'ATCG'\n",
    "        we append it to the empty string\n",
    "if then length of this new string is greater than one, then we have at least one char that is not ATCG.\n",
    "the below is not necesary for this question, but might be used in a following question.\n",
    "(then, we process the original string using the translate string method and the None type thing)\n",
    "\"\"\"\n",
    "# code\n",
    "\n",
    "# prompt user for DNA string\n",
    "dna_str1 = input(\"Please enter a DNA string that might contain non-ATCGs: \")\n",
    "# create an empty string to capture non-ATCG characters\n",
    "dna_str1_extra = \"\"\n",
    "\n",
    "for chr in dna_str1:\n",
    "    if chr not in \"ATCG\":\n",
    "        dna_str1_extra += chr\n",
    "    else:\n",
    "        pass\n",
    "if len(dna_str1_extra) > 0:\n",
    "    print(\"The original DNA string contains letters other than ATCGs.\")\n",
    "else:\n",
    "    print(\"The original DNA string only contains ATCGs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b5c2c62",
   "metadata": {},
   "source": [
    "2\\. Given that you have a string of comma separated values, like \"1,22,333,4,5\". How would you find the positions of the commas in the string?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d7c922a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please input a string of comma separated values, like 1,22,333,4,5: 1,22,333,4,5\n",
      "The positions of the commas in the string are [1, 4, 8, 10].\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "prompt user for a string of comma separated values, like 1,22,333,4,5.\n",
    "to find the comma positions, use a loop that is equal to range(len(str))\n",
    "initiate a list to record the positions of \",\".\n",
    "for position in range(len(str))\n",
    "    if str[position] == \",\"\n",
    "        list.append(position).\n",
    "    else, do nothing.\n",
    "the list records all the positions of \",\".\n",
    "\"\"\"\n",
    "# code\n",
    "usr_csv = input(\"Please input a string of comma separated values, like 1,22,333,4,5: \")\n",
    "# initiate empty list to track \",\" positions:\n",
    "comma_pos = []\n",
    "for position in range(len(usr_csv)):\n",
    "    if usr_csv[position] == \",\":\n",
    "        comma_pos.append(position)\n",
    "    else:\n",
    "        pass\n",
    "print(f\"The positions of the commas in the string are {comma_pos}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb0864de",
   "metadata": {},
   "source": [
    "3\\. In the files data1-4.gb find all titles and authors using stateful parsing. Put every author/title record (i.e. all authors/title in one string) in one list. Display all neatly with title on one line, authors on the next and a separator line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d466b1f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nucleotide sequence of a cDNA clone encoding human preproinsulin \n",
      "Bell,G.I., Swain,W.F., Pictet,R., Cordell,B., Goodman,H.M. and Rutter,W.J. \n",
      "\n",
      "Sequence of the human insulin gene \n",
      "Bell,G.I., Pictet,R.L., Rutter,W.J., Cordell,B., Tischer,E. and Goodman,H.M. \n",
      "\n",
      "Nucleotide sequence of human preproinsulin complementary DNA \n",
      "Sures,I., Goeddel,D.V., Gray,A. and Ullrich,A. \n",
      "\n",
      "Genetic variation in the human insulin gene \n",
      "Ullrich,A., Dull,T.J., Gray,A., Brosius,J. and Sures,I. \n",
      "\n",
      "The highly polymorphic region near the human insulin gene is composed of simple tandemly repeating sequences \n",
      "Bell,G.I., Selby,M.J. and Rutter,W.J. \n",
      "\n",
      "Variation in the sequence and modification state of the human insulin gene flanking regions \n",
      "Ullrich,A., Dull,T.J., Gray,A., Philips,J.A. III. and Peter,S. \n",
      "\n",
      "Analysis of the 3' end of the human pro-alpha 2(I) collagen gene. Utilization of multiple polyadenylation sites in cultured fibroblasts \n",
      "Myers,J.C., Dickson,L.A., de Wet,W.J., Bernard,M.P., Chu,M.L., Di Liberto,M., Pepe,G., Sangiorgi,F.O. and Ramirez,F. \n",
      "\n",
      "Analysis of the promoter region and the N-propeptide domain of the human pro alpha 2(I) collagen gene \n",
      "Dickson,L.A., de Wet,W., Di Liberto,M., Weil,D. and Ramirez,F. \n",
      "\n",
      "Organization of the human pro-alpha 2(I) collagen gene \n",
      "de Wet,W., Bernard,M., Benson-Chanda,V., Chu,M.L., Dickson,L., Weil,D. and Ramirez,F. \n",
      "\n",
      "Structure of a full-length cDNA clone for the prepro alpha 2(I) chain of human type I procollagen. Comparison with the chicken gene confirms unusual patterns of gene conservation \n",
      "Kuivaniemi,H., Tromp,G., Chu,M.L. and Prockop,D.J. \n",
      "\n",
      "Single base mutation in the pro alpha 2(I) collagen gene that causes efficient splicing of RNA from exon 27 to exon 29 and synthesis of a shortened but in-frame pro alpha 2(I) chain \n",
      "Tromp,G. and Prockop,D.J. \n",
      "\n",
      "A 19-base pair deletion in the pro-alpha 2(I) gene of type I procollagen that causes in-frame RNA splicing from exon 10 to exon 12 in a proband with atypical osteogenesis imperfecta and in his asymptomatic mother \n",
      "Kuivaniemi,H., Sabol,C., Tromp,G., Sippola-Thiele,M. and Prockop,D.J. \n",
      "\n",
      "Structural and functional analysis of the first intron of the human alpha2(I) collagen encoding gene \n",
      "Sherwood,A.L., Bottenus,R.E., Martzen,M.R. and Bornstein,P. \n",
      "\n",
      "Structural and functional analysis of the first intron of the human alpha2(I) collagen encoding gene \n",
      "Sherwood,A.L., Bottenus,R.E., Martzen,M.R. and Bornstein,P. \n",
      "\n",
      "Heterozygous mutation in the G+5 position of intron 33 of the pro-alpha 2(I) gene (COL1A2) that causes aberrant RNA splicing and lethal osteogenesis imperfecta. Use of carbodiimide methods that decrease the extent of DNA sequencing necessary to define an unusual mutation \n",
      "Ganguly,A., Baldwin,C.T., Strobel,D., Conway,D., Horton,W. and Prockop,D.J. \n",
      "\n",
      "Analysis of the COL1A1 and COL1A2 Genes by CSGE and DNA Sequencing in 14 Patients with Mild OI (Type I). Identification of Common Sequences for Null Allele Mutations \n",
      "Korkko,J.M., Earley,J.J., Ala-Kokko,L. and Prockop,D.J. \n",
      "\n",
      "Complete nucleotide sequence of intron 9 of the human pro-alpha2(I) collagen gene \n",
      "Trummer,T., Schwemmle,S., Just,W., Glaeser,D. and Kennerknecht,I. \n",
      "\n",
      "A highly polymorphic (ACT)n VNTR (variable nucleotide tandem repeats) locus inside intron 12 of COL1A2, one of the two genes involved in dominant osteogenesis imperfecta \n",
      "Pepe,G. \n",
      "\n",
      "Direct Submission \n",
      "Korkko,J.M., Earley,J.J., De Paepe,A., Nuytinck,L., Ala-Kokko,L. and Prockop,D.J. \n",
      "\n",
      "Structural organization and sequence analysis of the globin locus in Atlantic salmon \n",
      "McMorrow,T., Wagner,A., Deryckere,F. and Gannon,F. \n",
      "\n",
      "Direct Submission \n",
      "McMorrow,T. \n",
      "\n",
      "Cloning and characterisation of the gene encoding mole rat (Spalax ehrenbergi) growth hormone \n",
      "Lioupis,A., Nevo,E. and Wallis,M. \n",
      "\n",
      "Direct Submission \n",
      "Lioupis,A. \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "red flag is when tile line has ended, i.e. JOURNAL starts\n",
    "we want the contents that comes after AUTHOR, and before JOURNAL, putting every author/title record in one string, in one list.\n",
    "green flag is when AUTHORS line has started\n",
    "* remember str.startswith method is very helpful.\n",
    "had to place the order this way:\n",
    "# green flag\n",
    "# red flag\n",
    "# data gathering.\n",
    "\n",
    "Then, needed to use.strip() + \" \" while collecting the contents of each lines, so that it display correctly later.\n",
    "Use .replace() twice to later correctly splice the string into a single list that to be displayed :).\n",
    "\n",
    "Use a more specific method instead of the a,b = b,a to display the result correctly.\n",
    "\n",
    "\"\"\"\n",
    "# code\n",
    "\n",
    "# prompt user to enter the file name.\n",
    "def main():\n",
    "    # using one file first, will change into all files later.\n",
    "    data_files = [\"data1.gb\", \"data2.gb\", \"data3.gb\", \"data4.gb\"]\n",
    "    for data_file in data_files:\n",
    "        display_auth_title(data_file)\n",
    "\n",
    "def display_auth_title(file):\n",
    "    with open (f\"{file}\", \"r\") as infile:\n",
    "    # initiat stateful parsing with an empty author_title string, with the flag set to false.\n",
    "        (auth_title_v1, auth_title_flag) = (\"\", False)\n",
    "        # The below flag positons allows the words \"AUTHORS\", \"TITLE\" to remain in the final string.\n",
    "        for line in infile:\n",
    "            if line.strip().startswith(\"AUTHORS\"):    # green flag\n",
    "                auth_title_flag = True\n",
    "            if line.strip().startswith(\"JOURNAL\"):    # red flag\n",
    "                auth_title_flag = False\n",
    "            if auth_title_flag:\n",
    "                #.strip()+\" \" makes sure places that needed a space have it inplace correctly\n",
    "                # this works better than .replace(\"\\n\". \" \")\n",
    "                auth_title_v1 += line[:].strip()+\" \"    \n",
    "        if auth_title_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if auth_title_v1 == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    # splite the string using two delimiters \n",
    "    auth_title_v2 = auth_title_v1.replace(\"AUTHORS   \", \"%\").replace(\"TITLE     \", \"%\")\n",
    "    auth_title_list1 = auth_title_v2.split(\"%\")\n",
    "    auth_title_list1.pop(0)\n",
    "    # get rid of the empty space item at the start:\n",
    "    # the auth_title_list1 items are authors, title, authors, title\n",
    "    # first switch the positions of the items\n",
    "    # to have title infront of author.\n",
    "    # method below to switch positions worked best. The usual a,b = b,a ended up with incorrect items at the start or end.\n",
    "    new_positions = []\n",
    "    for position in range(len(auth_title_list1)):\n",
    "        if position % 2 == 1:\n",
    "            new_positions.append(position)\n",
    "            new_positions.append(position-1)\n",
    "\n",
    "    for position in new_positions:\n",
    "        print(auth_title_list1[position])       # prints title then author\n",
    "        if position % 2 == 0:                   # if we are at the end of an author\n",
    "            print()                             # prints separator line\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d4ee15b",
   "metadata": {},
   "source": [
    "4\\. Continue with exercise 4: Your program already extracts the DNA. Make it also extract the positions of the coding sequence. Display them. You can recognize it by the line that starts with \"     CDS     join(\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0302fe60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a file Genbank entry file, such as one of data1-4.gb: data2.gb\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline, translated, DNA, CDS positions (STOP to quit search): CDS positions\n",
      "\n",
      "Below are the CDS positions: \n",
      "2432..2501,5144..5154,5746..5760,6411..6446,7558..7650,8929..8982,11919..11963,12056..12109,12202..12255,12562..12615,13036..13089,13613..13666,15217..15261,15553..15606,15706..15750,16140..16193,16692..16790,16934..16978,17093..17191,17612..17665,17791..17898,18259..18312,18426..18524,19436..19489,19953..20051,20452..20505,21059..21112,21263..21316,21590..21643,22596..22640,23773..23871,25090..25197,25867..25920,26854..26907,27588..27641,27746..27799,27896..28003,28365..28418,29217..29270,30273..30434,31647..31754,32427..32534,32920..32973,33060..33167,33308..33361,33733..33840,34316..34369,34496..34603,34936..35194,35602..35786,36498..36740,37554..37700\n",
      "\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline, translated, DNA, CDS positions (STOP to quit search): STOP\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "The code below is copied from ex4\n",
    "create new function that extracts the position of the coding sequence.\n",
    "edit the user prompt so they can ask for the CDS positions.\n",
    "\"\"\"\n",
    "\n",
    "# code\n",
    "def main():\n",
    "    usr_filename = input(\"Please enter a file Genbank entry file, such as one of data1-4.gb: \")\n",
    "    find_keyword(usr_filename)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def find_keyword(filename):\n",
    "# make a loop so usr can keep prompting and we can keep displaying\n",
    "    while True:\n",
    "        search_word = input(\"Specify what you want to know by typing one of the following:\\naccession number, definition, organism, medline, translated, DNA, CDS positions (STOP to quit search): \")\n",
    "        if search_word == \"medline\":\n",
    "            find_medline(filename)\n",
    "        elif search_word in (\"accession number\", \"definition\", \"organism\"):\n",
    "            find_an_df_og(filename, search_word)    # very important to include \"search_word\" here.\n",
    "        elif search_word == \"translated\":\n",
    "            find_translated(filename)\n",
    "        elif search_word == \"DNA\":\n",
    "            find_DNA(filename)\n",
    "        elif search_word == \"CDS positions\":\n",
    "            find_CDS(filename)\n",
    "        elif search_word == \"STOP\":\n",
    "            break\n",
    "        else:\n",
    "            print(\"I don't think we have what you typed, let's try again.\")\n",
    "            pass\n",
    "\n",
    "\n",
    "\n",
    "def find_CDS(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, cds_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"CDS             join(\"):    # green flag\n",
    "                cds_flag = True\n",
    "            if cds_flag:\n",
    "                result += line\n",
    "            if line.endswith(\")\"):                          # red flag\n",
    "                cds_flag = False\n",
    "        if cds_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(\"\\nBelow are the CDS positions: \")\n",
    "    for char in result:\n",
    "        if not char.isalpha():\n",
    "            char = char.strip(\"(\").strip(\")\").strip()\n",
    "            print(char, end = \"\")\n",
    "    print()\n",
    "    print()\n",
    "\n",
    "def find_DNA(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, dna_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"//\"):        # red flag\n",
    "                dna_flag = False\n",
    "            if dna_flag:\n",
    "                result += ''.join(line.split()[1:])\n",
    "            if line.startswith(\"ORIGIN\"):    # green flag\n",
    "                dna_flag = True\n",
    "        if dna_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(f\"\\nThe DNA sequence of the translated gene is:\\n{result}\\n\")\n",
    "\n",
    "def find_translated(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, tran_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            # this is the red line\n",
    "            # put the red line at the top so it is not included in the result :)\n",
    "            # this way the results stop at the correct place.\n",
    "            if line.startswith(\"exon\"):\n",
    "                tran_flag = False\n",
    "            if tran_flag:\n",
    "                # carful here, the keyword in the file is translation, not translated :).\n",
    "                result += line.strip(\"/translation=\").strip('\"').strip()\n",
    "            # this is the green line\n",
    "            # carful here with, remember the correct word is translation.\n",
    "            if line.startswith(\"/translation=\"):\n",
    "                tran_flag = True\n",
    "        if tran_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(f\"\\nThe DNA sequence of the translated gene is:\\n{result}\\n\")\n",
    "\n",
    "\n",
    "def find_medline(filename):\n",
    "    medlines = []\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"MEDLINE\"):\n",
    "                medlines.append(line.strip(\"MEDLINE\").strip())\n",
    "            else:\n",
    "                pass\n",
    "    if len(medlines) > 1:\n",
    "        are = \"are\"\n",
    "    else:\n",
    "        are = \"is\"\n",
    "    print(f\"The medlines {are}: {medlines}\")\n",
    "\n",
    "\n",
    "def find_an_df_og(filename, search_word):\n",
    "    with open (f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # line.strip() will help with key word search like ORGANISM line, which starts with two spaces.\n",
    "            line = line.strip()\n",
    "            # the find start parts can be made into a smaller programme, but not that fit for what ex1 asks for.\n",
    "            # the way to do this is to ask the user for a string, that equals to \"ACCESSION\", \"DEFINITION\", ...etc\n",
    "            # search_word.split()[0] is useful for cases when we write accession number.\n",
    "            if line.startswith(f\"{search_word.split()[0].upper()}\"):\n",
    "                result = line.strip(f\"{search_word.split()[0].upper()}\").strip()\n",
    "                print(f\"The {search_word} is {result}.\")\n",
    "            elif len(search_word) == 0:\n",
    "                print(\"Empty search term!\")\n",
    "            else:\n",
    "                # using pass here for now\n",
    "                pass\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4e7195d",
   "metadata": {},
   "source": [
    "5\\. Continue with exercise 4: Now that you have the DNA and the positions in the DNA for the coding sequence, extract the coding DNA directly from the DNA, such that you can display the actual gene. Verify that you have done so correctly by thinking about the properties of coding DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0e9f8b64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a file Genbank entry file, such as one of data1-4.gb: data2.gb\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline, translated, DNA, CDS positions, Coding DNA(STOP to quit search): Coding DNA\n",
      "The DNA sequences displayed below are the actual coding DNA:\n",
      "atgctcagctttgtggatacgcggactttgttgctgcttgcagtaaccttatgcctagcaacatgccaatg\n",
      "\n",
      "ctttacaagagg\n",
      "\n",
      "gaaactgtaagaaagg\n",
      "\n",
      "ggcccagccggagatagaggaccacgtggagaaaggg\n",
      "\n",
      "ggtccaccaggccccccaggcagagatggtgaagatggtcccacaggccctcctggtccacctggtcctcctggcccccctggtctcggtgggg\n",
      "\n",
      "aactttgctgctcagtatgatggaaaaggagttggacttggccctggaccaatgg\n",
      "\n",
      "ggcttaatgggacctagaggcccacctggtgcagctggagccccag\n",
      "\n",
      "ggccctcaaggtttccaaggacctgctggtgagcctggtgaacctggtcaaactg\n",
      "\n",
      "ggtcctgcaggtgctcgtggtccagctggccctcctggcaaggctggtgaagatg\n",
      "\n",
      "ggtcaccctggaaaacccggacgacctggtgagagaggagttgttggaccacagg\n",
      "\n",
      "ggtgctcgtggtttccctggaactcctggacttcctggcttcaaaggcattaggg\n",
      "\n",
      "ggacacaatggtctggatggattgaagggacagcccggtgctcctggtgtgaagg\n",
      "\n",
      "ggtgaacctggtgcccctggtgaaaatggaactccaggtcaaacag\n",
      "\n",
      "ggagcccgtgggcttcctggtgagagaggacgtgttggtgcccctggcccagctg\n",
      "\n",
      "ggtgcccgtggcagtgatggaagtgtgggtcccgtgggtcctgctg\n",
      "\n",
      "ggtcccattgggtctgctggccctccaggcttcccaggtgcccctggccccaagg\n",
      "\n",
      "ggtgaaattggagctattggtaacgctggtcctgctggtcccgccggtccccgtggtgaagtgggtcttccaggcctctccggccccgttggacctcctg\n",
      "\n",
      "ggtaatcctggagcaaacggccttactggtgccaagggtgctgctg\n",
      "\n",
      "ggccttcccggcgttgctggggctcccggcctccctggaccccgcggtattcctggccctgttggtgctgccggtgctactggtgccagaggacttgttg\n",
      "\n",
      "ggtgagcctggtccagctggctccaaaggagagagcggtaacaagggtgagcccg\n",
      "\n",
      "ggctctgctgggccccaaggtcctcctggtcccagtggtgaagaaggaaagagaggccctaatggggaagctggatctgccggccctccaggacctcctgggctgagag\n",
      "\n",
      "ggtagtcctggttctcgtggtcttcctggagctgatggcagagctggcgtcatgg\n",
      "\n",
      "ggccctcctggtagtcgtggtgcaagtggccctgctggagtccgaggacctaatggagatgctggtcgccctggggagcctggtctcatgggacccagag\n",
      "\n",
      "ggtcttcctggttcccctggaaatatcggccccgctggaaaagaaggtcctgtcg\n",
      "\n",
      "ggcctccctggcatcgacggcaggcctggcccaattggccccgttggagcaagaggagagcctggcaacattggattccctggacccaaaggccccactg\n",
      "\n",
      "ggtgaccctggcaaaaacggtgataaaggtcatgctggtcttgctggtgctcggg\n",
      "\n",
      "ggtgctccaggtcctgatggaaacaatggtgctcagggacctcctggaccacagg\n",
      "\n",
      "ggtgttcaaggtggaaaaggtgaacagggtcccgctggtcctccaggcttccagg\n",
      "\n",
      "ggtctgcctggcccctcaggtcccgctggtgaagttggcaaaccaggagaaaggg\n",
      "\n",
      "ggtctccatggtgagtttggtctccctggtcctgctggtccaagag\n",
      "\n",
      "ggggaacgcggtcccccaggtgagagtggtgctgccggtcctactggtcctattggaagccgaggtccttctggacccccagggcctgatggaaacaagg\n",
      "\n",
      "ggtgaacctggtgtggttggtgctgtgggcactgctggtccatctggtcctagtggactcccaggagagaggggtgctgctggcatacctggaggcaagggagaaaagg\n",
      "\n",
      "ggtgaacctggtctcagaggtgaaattggtaaccctggcagagatggtgctcgtg\n",
      "\n",
      "ggtgctcatggtgctgtaggtgcccctggtcctgctggagccacaggtgaccggg\n",
      "\n",
      "ggcgaagctggggctgctggtcctgctggtcctgctggtcctcggggaagccctg\n",
      "\n",
      "ggtgaacgtggcgaggtcggtcctgctggccccaacggatttgctggtccggctg\n",
      "\n",
      "ggtgctgctggtcaaccgggtgctaaaggagaaagaggaggcaaagggcctaagggtgaaaacggtgttgttggtcccacaggccccgttggagctgctggcccagctg\n",
      "\n",
      "ggtccaaatggtccccccggtcctgctggaagtcgtggtgatggaggcccccctg\n",
      "\n",
      "ggtatgactggtttccctggtgctgctggacggactggtcccccaggaccctctg\n",
      "\n",
      "ggtatttctggccctcctggtccccctggtcctgctgggaaagaagggcttcgtggtcctcgtggtgaccaaggtccagttggccgaactggagaagtaggtgcagttggtccccctggcttcgctggtgagaagggtccctctggagaggctggtactgctg\n",
      "\n",
      "ggacctcctggcactccaggtcctcagggtcttcttggtgctcctggtattctgggtctccctggctcgagaggtgaacgtggtctacctggtgttgctggtgctgtgg\n",
      "\n",
      "ggtgaacctggtcctcttggcattgccggccctcctggggcccgtggtcctcctggtgctgtgggtagtcctggagtcaacggtgctcctggtgaagctggtcgtgatg\n",
      "\n",
      "ggcaaccctgggaacgatggtcccccaggtcgcgatggtcaacccggacacaagg\n",
      "\n",
      "ggagagcgcggttaccctggcaatattggtcccgttggtgctgcaggtgcacctggtcctcatggccccgtgggtcctgctggcaaacatggaaaccgtggtgaaactg\n",
      "\n",
      "ggtccttctggtcctgttggtcctgctggtgctgttggcccaagaggtcctagtg\n",
      "\n",
      "ggcccacaaggcattcgtggcgataagggagagcccggtgaaaaggggcccagaggtcttcctggcttcaagggacacaatggattgcaaggtctgcctggtatcgctg\n",
      "\n",
      "ggtcaccatggtgatcaaggtgctcctggctccgtgggtcctgctggtcctaggg\n",
      "\n",
      "ggccctgctggtccttctggccctgctggaaaagatggtcgcactggacatcctggtacggttggacctgctggcattcgaggccctcagggtcaccaaggccctgctg\n",
      "\n",
      "ggcccccctggtccccctggccctcctggacctccaggtgtaagcggtggtggttatgactttggttacgatggagacttctacagggctgaccagcctcgctcagcaccttctctcagacccaaggactatgaagttgatgctactctgaagtctctcaacaaccagattgagacccttcttactcctgaaggctctagaaagaacccagctcgcacatgccgtgacttgagactcagccacccagagtggagcagtgg\n",
      "\n",
      "gttactactggattgaccccaaccaaggatgcactatggaagccatcaaagtatactgtgatttccctaccggcgaaacctgtatccgggcccaacctgaaaacatcccagccaagaactggtataggagctccaaggacaagaaacacgtctggctaggagaaactatcaatgctggcagccagg\n",
      "\n",
      "tttgaatataatgttgaaggagtgacttccaaggaaatggctacccaacttgccttcatgcgcctgctggccaactatgcctctcagaacatcacctaccactgcaagaacagcattgcatacatggatgaggagactggcaacctgaaaaaggctgtcattctacagggctctaatgatgttgaacttgttgctgagggcaacagcaggttcacttacactgttcttgtagatggctgctctg\n",
      "\n",
      "aaaaagacaaatgaatggggaaagacaatcattgaatacaaaacaaataagccatcacgcctgcccttccttgatattgcacctttggacatcggtggtgctgaccatgaattctttgtggacattggcccagtctgtttcaaataaa\n",
      "\n",
      "Specify what you want to know by typing one of the following:\n",
      "accession number, definition, organism, medline, translated, DNA, CDS positions, Coding DNA(STOP to quit search): STOP\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "The code below is copied from L10_ex4_extra.py\n",
    "edit so that the user can search for \"coding DNA\"\n",
    "create a new function definition that does a bit of both from the function for find_DNA and find_CDS\n",
    "display the correct parts of the coding DNA sequence.\n",
    "\"\"\"\n",
    "\n",
    "# pseudo code\n",
    "\"\"\"\n",
    "The code below is copied from ex4\n",
    "create new function that extracts the position of the coding sequence.\n",
    "edit the user prompt so they can ask for the CDS positions.\n",
    "\"\"\"\n",
    "\n",
    "# code\n",
    "def main():\n",
    "    usr_filename = input(\"Please enter a file Genbank entry file, such as one of data1-4.gb: \")\n",
    "    find_keyword(usr_filename)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def find_keyword(filename):\n",
    "# make a loop so usr can keep prompting and we can keep displaying\n",
    "    while True:\n",
    "        search_word = input(\"Specify what you want to know by typing one of the following:\\naccession number, definition, organism, medline, translated, DNA, CDS positions, Coding DNA(STOP to quit search): \")\n",
    "        if search_word == \"medline\":\n",
    "            find_medline(filename)\n",
    "        elif search_word in (\"accession number\", \"definition\", \"organism\"):\n",
    "            find_an_df_og(filename, search_word)    # very important to include \"search_word\" here.\n",
    "        elif search_word == \"translated\":\n",
    "            find_translated(filename)\n",
    "        elif search_word == \"DNA\":\n",
    "            find_DNA(filename)\n",
    "        elif search_word == \"CDS positions\":\n",
    "            find_CDS(filename)\n",
    "        elif search_word == \"Coding DNA\":\n",
    "            find_cDNA(filename)\n",
    "        elif search_word == \"STOP\":\n",
    "            break\n",
    "        else:\n",
    "            print(\"I don't think we have what you typed, let's try again.\")\n",
    "            pass\n",
    "\n",
    "\n",
    "def find_cDNA(filename):\n",
    "    print(\"The DNA sequences displayed below are the actual coding DNA:\")\n",
    "    list_of_positions = CDS_position_list(filename)\n",
    "    str_dna_seq = DNA_seq_list(filename)\n",
    "    for item in list_of_positions:\n",
    "        start_of = int(item.split(\"..\")[0])-1\n",
    "        end_of = int(item.split(\"..\")[1])+1\n",
    "        print(str_dna_seq[start_of:end_of])\n",
    "        print()                                 # separator sequence\n",
    "\n",
    "def DNA_seq_list(filename):\n",
    "    # first part if very similar to the function for find_DNA.\n",
    "    # main difference is that the result is saved to one list\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, dna_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"//\"):        # red flag\n",
    "                dna_flag = False\n",
    "            if dna_flag:\n",
    "                result += ''.join(line.split()[1:])\n",
    "            if line.startswith(\"ORIGIN\"):    # green flag\n",
    "                dna_flag = True\n",
    "        if dna_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    return result\n",
    "\n",
    "def CDS_position_list(filename):\n",
    "    # first part is very similar to the function for find_CDS.\n",
    "    # main difference is that the result is saved to one list\n",
    "    # each item being one CDS region.\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, cds_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"CDS             join(\"):    # green flag\n",
    "                cds_flag = True\n",
    "            if cds_flag:\n",
    "                result += line\n",
    "            if line.endswith(\")\"):                          # red flag\n",
    "                cds_flag = False\n",
    "        if cds_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"Not data found. Can't trust result.\")\n",
    "    CDS_str1 = \"\"\n",
    "    for char in result:\n",
    "        if not char.isalpha():\n",
    "            char = char.strip(\"(\").strip(\")\").strip()\n",
    "            CDS_str1+= char\n",
    "    CDS_list1 = CDS_str1.split(\",\")\n",
    "    return CDS_list1     # list looks like this ['203..295', '522..743', '877..1005']\n",
    "\n",
    "\n",
    "def find_CDS(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, cds_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"CDS             join(\"):    # green flag\n",
    "                cds_flag = True\n",
    "            if cds_flag:\n",
    "                result += line\n",
    "            if line.endswith(\")\"):                          # red flag\n",
    "                cds_flag = False\n",
    "        if cds_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(\"\\nBelow are the CDS positions: \")\n",
    "    for char in result:\n",
    "        if not char.isalpha():\n",
    "            char = char.strip(\"(\").strip(\")\").strip()\n",
    "            print(char, end = \"\")\n",
    "    print()\n",
    "    print()\n",
    "\n",
    "def find_DNA(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, dna_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"//\"):        # red flag\n",
    "                dna_flag = False\n",
    "            if dna_flag:\n",
    "                result += ''.join(line.split()[1:])\n",
    "            if line.startswith(\"ORIGIN\"):    # green flag\n",
    "                dna_flag = True\n",
    "        if dna_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(f\"\\nThe DNA sequence of the translated gene is:\\n{result}\\n\")\n",
    "\n",
    "def find_translated(filename):\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        (result, tran_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            # this is the red line\n",
    "            # put the red line at the top so it is not included in the result :)\n",
    "            # this way the results stop at the correct place.\n",
    "            if line.startswith(\"exon\"):\n",
    "                tran_flag = False\n",
    "            if tran_flag:\n",
    "                # carful here, the keyword in the file is translation, not translated :).\n",
    "                result += line.strip(\"/translation=\").strip('\"').strip()\n",
    "            # this is the green line\n",
    "            # carful here with, remember the correct word is translation.\n",
    "            if line.startswith(\"/translation=\"):\n",
    "                tran_flag = True\n",
    "        if tran_flag:\n",
    "            raise ValueError(\"Format is not right. Can't trust result.\")\n",
    "        if result == \"\":\n",
    "            raise ValueError(\"No data found. Can't trust result.\")\n",
    "    print(f\"\\nThe DNA sequence of the translated gene is:\\n{result}\\n\")\n",
    "\n",
    "\n",
    "def find_medline(filename):\n",
    "    medlines = []\n",
    "    with open(f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\"MEDLINE\"):\n",
    "                medlines.append(line.strip(\"MEDLINE\").strip())\n",
    "            else:\n",
    "                pass\n",
    "    if len(medlines) > 1:\n",
    "        are = \"are\"\n",
    "    else:\n",
    "        are = \"is\"\n",
    "    print(f\"The medlines {are}: {medlines}\")\n",
    "\n",
    "\n",
    "def find_an_df_og(filename, search_word):\n",
    "    with open (f\"{filename}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            # line.strip() will help with key word search like ORGANISM line, which starts with two spaces.\n",
    "            line = line.strip()\n",
    "            # the find start parts can be made into a smaller programme, but not that fit for what ex1 asks for.\n",
    "            # the way to do this is to ask the user for a string, that equals to \"ACCESSION\", \"DEFINITION\", ...etc\n",
    "            # search_word.split()[0] is useful for cases when we write accession number.\n",
    "            if line.startswith(f\"{search_word.split()[0].upper()}\"):\n",
    "                result = line.strip(f\"{search_word.split()[0].upper()}\").strip()\n",
    "                print(f\"The {search_word} is {result}.\")\n",
    "            elif len(search_word) == 0:\n",
    "                print(\"Empty search term!\")\n",
    "            else:\n",
    "                # using pass here for now\n",
    "                pass\n",
    "\n",
    "\n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
