{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "60beea57",
   "metadata": {},
   "source": [
    "## L12: Dictionaries\n",
    "Exercises for Extra Practice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acb690a0",
   "metadata": {},
   "source": [
    "1\\. Given a tab-separated file with 3 columns; StudentID, CourseNumber, Grade. Can you find a way to load the grades for a student in a retrievable manner into (some of) the python data structures learned so far? Retrievable means here that you can find the grades for a student if you know the studentID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "53b60664",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the studentID and the associated grade will be displayed (STOP to exit): 215047\n",
      "5\n",
      "Enter the studentID and the associated grade will be displayed (STOP to exit): 210554\n",
      "9\n",
      "Enter the studentID and the associated grade will be displayed (STOP to exit): 221660\n",
      "4\n",
      "Enter the studentID and the associated grade will be displayed (STOP to exit): STOP\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "Given tab-separated file with three columns: StudentID, CourseNumber, Grade.\n",
    "Load the grades for a student into a dictionary\n",
    "So that giveb the student ID, I can find the grades.\n",
    "Explain your approach. Hint: It is not necessarily efficient.\n",
    "\n",
    "\n",
    "Made a toy tab-seaprated file to work on\n",
    "\n",
    "def main():\n",
    "    variable = input(prompt user to ente studentID:)\n",
    "    return_grades(studentID)\n",
    "\n",
    "def return_grades(id):\n",
    "    create empty dictionary id_grade_dict\n",
    "    with open file:\n",
    "        for line in file:\n",
    "            id_grade_dict[line[0]] = line[-1]\n",
    "    # when writting the code, print out the dictionary as a sanity check\n",
    "    \n",
    "    #then let's get back to this function:\n",
    "    return id_grade_dict[id]\n",
    "    # maybe print a message here.\n",
    "\n",
    "main\n",
    "\"\"\"\n",
    "# code\n",
    "def main():\n",
    "    while True:\n",
    "        studentID = input(\"Enter the studentID and the associated grade will be displayed (STOP to exit): \")\n",
    "        if studentID != \"STOP\":\n",
    "            return_grades(studentID)\n",
    "        else:\n",
    "            break\n",
    "\n",
    "def return_grades(id):\n",
    "    id_grade_dict = dict()\n",
    "    with open(\"students.txt\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.split()\n",
    "            if line[0].isnumeric():\n",
    "                id_grade_dict[line[0]] = line[-1]\n",
    "    print(id_grade_dict[f\"{id}\"])\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bf120a3",
   "metadata": {},
   "source": [
    "2\\. The geneA-E.txt files all have the same structure on each line; first number is a float between 0 and 1, second number is an integer. For all files (the combined data set) find the average of the float, given the integer and display in ascending order of the integer. You need to add all the floats for a given integer together and divide by the number of floats for the integer, then you have the average for the integer. To succeed at this, you must use two dicts where the integer is the key in both. The corresponding values are the sum of the floats (for that key) and the number of times the key has been encountered in the files.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f929c843",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter all file names, separated by space: geneA.txt geneB.txt geneC.txt geneD.txt geneE.txt\n",
      "Please enter the integer for which you want the float avgs of (STOP to exit): 16\n",
      "The average float value for 16 is 0.1563.\n",
      "Please enter the integer for which you want the float avgs of (STOP to exit): 90\n",
      "The average float value for 90 is 0.1856.\n",
      "Please enter the integer for which you want the float avgs of (STOP to exit): STOP\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "create a combined data set for geneA-E.txt\n",
    "then create one dict as described above\n",
    "\n",
    "code would look something like:\n",
    "can add a loop into main so user can keep asking again.\n",
    "def main():\n",
    "    usr_file_list = [input(Prompt for user to enter all file names, separated by space)]\n",
    "    usr_int = int(input(Prompt user to input an integer))\n",
    "    # first, to make a file with all data combined.\n",
    "    urs_dict_sum, usr_dict_freq = sum_freq_data(usr_file_list)\n",
    "    display_avg_float(usr_dict_sum, usr_dict_freq, usr_int)\n",
    "\n",
    "\n",
    "# this function takes in two dictionaries, one with the sum of the values, one with the key's occurence freq.\n",
    "def display_avg_float(dict_sum, dict_freq, int_key):\n",
    "    avg_float = dict_sum[int_key] / dict_freq[int_key]\n",
    "    print(avg_float)\n",
    "\n",
    "\n",
    "# this function takes a list of files as input\n",
    "# returns two dictionaries as output, both with the same set of int keys.\n",
    "    # one dict has the values as the sum of the floats\n",
    "    # one dict counts the number of times a key has occured.\n",
    "# these two dicts are then used as input for finding avg.\n",
    "def sum_freq_data(file_list):\n",
    "    dict_sum = dict{}\n",
    "    dict_freq = dict{}\n",
    "    for file in file_list:\n",
    "        with open(f\"{file}\", \"r\") as infile:\n",
    "            for line in infile:\n",
    "                # hope this works\n",
    "                if column value not in list of dict keys:\n",
    "                    dict_sum[col2] = col1\n",
    "                    dict_freq[col2] = 1    # i.e. first occurence.\n",
    "                else:\n",
    "                    dict_sum[col2] += col1\n",
    "                    dict_freq[col2] += 1\n",
    "\n",
    "    return (dict_sum, dict_freq)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "# code\n",
    "\n",
    "def main():\n",
    "    usr_file_list = input(\"Please enter all file names, separated by space: \").split()\n",
    "    while True:\n",
    "        # usr_file_list is the input for sum_freq_data().\n",
    "        # sum_freq_data() returns two dics:\n",
    "        #     one with the sums of all floats for the same key.\n",
    "        #    one with the number of occurences of the same key.\n",
    "        usr_dict_sum, usr_dict_freq = sum_freq_data(usr_file_list)\n",
    "        usr_int = input(\"Please enter the integer for which you want the float avgs of (STOP to exit): \")\n",
    "        if usr_int != \"STOP\":\n",
    "            display_avg_float(usr_dict_sum, usr_dict_freq, int(usr_int))\n",
    "        else:\n",
    "            break\n",
    "\n",
    "def display_avg_float(dict_sum, dict_freq, int_key):\n",
    "    avg_float = dict_sum[int_key]/dict_freq[int_key]\n",
    "    avg_float = \"{:.4f}\".format(round(avg_float, 4))\n",
    "    print(f\"The average float value for {int_key} is {avg_float}.\")\n",
    "    # add print message to display.\n",
    "\n",
    "# remember the first column is float, the second column is int.\n",
    "# this function takes a list of files as input\n",
    "# returns two dictionaries as output, both with the same set of int keys.\n",
    "    # one dict has the values as the sum of the floats\n",
    "    # one dict counts the number of times a key has occured.\n",
    "# these two dicts are then used as input for finding avg.\n",
    "def sum_freq_data(file_list):\n",
    "    dict_sum = dict()\n",
    "    dict_freq = dict()\n",
    "    for file in file_list:\n",
    "        with open(f\"{file}\", \"r\") as infile:\n",
    "            for line in infile:\n",
    "                line = line.split()\n",
    "                col1 = float(line[0])\n",
    "                col2 = int(line[1])\n",
    "                if col2 not in list(dict_sum.keys()):\n",
    "                    dict_sum[col2] = col1\n",
    "                    dict_freq[col2] = 1    # first occurence\n",
    "                else:\n",
    "                    dict_sum[col2] += col1\n",
    "                    dict_freq[col2] += 1\n",
    "    return(dict_sum, dict_freq)\n",
    "\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a49a5f",
   "metadata": {},
   "source": [
    "3\\. This exercise requires that you did the last two practice exercises in Simple Pattern Matching. In the data1-4.gb files count who many times the different codons in the coding sequence occurs. Display."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5ea721d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'atg': 10, 'ctc': 14, 'agc': 9, 'ttt': 7, 'gtg': 14, 'gat': 28, 'acg': 3, 'cgg': 5, 'act': 27, 'ttg': 7, 'ctg': 13, 'ctt': 23, 'gca': 15, 'gta': 7, 'acc': 6, 'tta': 2, 'tgc': 5, 'cta': 6, 'aca': 11, 'caa': 15, 't': 1, 'tac': 8, 'aag': 30, 'ag': 2, 'gaa': 41, 'aga': 24, 'ggc': 87, 'cca': 45, 'gcc': 21, 'gga': 84, 'cgt': 20, 'agg': 9, 'ggt': 193, 'ccc': 49, 'cct': 137, 'ggg': 18, 'aac': 25, 'gct': 92, 'cag': 14, 'tat': 6, 'aaa': 22, 'gtt': 28, 'ttc': 15, 'gag': 22, 'cac': 8, 'cga': 5, 'att': 22, 'aat': 14, 'agt': 12, 'tct': 18, 'tcc': 8, 'cgc': 10, 'gtc': 7, 'atc': 6, 'gac': 12, 'cat': 8, 'tca': 5, 'ata': 2, 'ccg': 3, 'tcg': 1, 'tgg': 3, 'g': 1, 'gcg': 1, 'tag': 1, 'tgt': 1, 'taa': 1}\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "def main()\n",
    "for the list of files, do below for each file:\n",
    "how I would do it for one file:\n",
    "    input one file into get_cds_positions(), which will return one list of cds positions\n",
    "    input one file into get_seq(), which will return one list of the dna seq.\n",
    "    input the reutrn values of the above, i.e. the list of cds positions, and the dna seq to get_cds(),\n",
    "        which will return a list of actual cds, with the associated dna.\n",
    "    input one list of cds into codon_freq(), will return a dictionary, which counts the occurence of different codons.\n",
    "\n",
    "what does get_cds_positions() do?\n",
    "use simple pattern matching to extrace the coding positions from a file into one list.\n",
    "\n",
    "what does get_seq() do?\n",
    "use simple pattern matching to extract the cds\n",
    "\n",
    "what does get_cds() do?\n",
    "given a list of cds positions, and a string of a dna sequence\n",
    "returns a list of dna sequences, with each string being the coding sequence.\n",
    "\n",
    "\n",
    "what does codon_freq() do?\n",
    "initiate an empty codon_freq_dict\n",
    "makes a list that is equal to the three stop codon strings.\n",
    "loop through the codon list, making sure to slice it correctly:\n",
    "    if these three are not in the list(codon_freq_dict.key()):\n",
    "        codon_freq_dict[these three] = 1    # ininitate count.\n",
    "    elif these three are equal to the stop codons:\n",
    "        break\n",
    "    else these three are already in the list(codon_freq_dict.key()):\n",
    "        codon_freq_dict[these three] += 1\n",
    "at the end of the loop, return a dictionary codon_freq_dict.\n",
    "\"\"\"\n",
    "# code\n",
    "def main():\n",
    "    usr_positions = get_cds_positions(\"data2.gb\")\n",
    "    usr_dna_seq = get_dna_seq(\"data2.gb\")\n",
    "    usr_dna_cds = get_cds(usr_positions, usr_dna_seq)\n",
    "    print(codon_counts(usr_dna_cds))\n",
    "\n",
    "\n",
    "\n",
    "# recieves a list of aa sequences used for coding\n",
    "# returns the counts of occurences of different codons.\n",
    "def codon_counts(dna_cds):\n",
    "    # inidiate empty counter dict:\n",
    "    codon_counts = {}\n",
    "    for item in dna_cds:\n",
    "        for position in range(0,len(item),3):\n",
    "            codon = item[position:position+3]\n",
    "            # adjust to the new correct start of the reading frame.\n",
    "            if codon in codon_counts:\n",
    "                codon_counts[codon] += 1\n",
    "            else:\n",
    "                codon_counts[codon] = 1\n",
    "    return codon_counts\n",
    "\n",
    "\n",
    "def get_cds(positions, seq):\n",
    "    cds = []\n",
    "    for position in positions:\n",
    "        start_position = int(position.split(\"..\")[0]) - 1 # use -1 to account for zero indexing.\n",
    "        end_position = int(position.split(\"..\")[1])\n",
    "        cds.append(seq[start_position:end_position])\n",
    "    return cds\n",
    "\n",
    "\n",
    "\n",
    "def get_dna_seq(file):\n",
    "    with open(f\"{file}\", \"r\") as infile:\n",
    "        (seq, seq_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            if line.startswith(\"//\"): # Red Flag\n",
    "                seq_flag = False\n",
    "            if seq_flag:\n",
    "                # [1:] will exclude the leadning numbers.\n",
    "                seq += \"\".join(line.lstrip().split()[1:])\n",
    "            if line.startswith(\"ORIGIN\"): # Green Flag\n",
    "                seq_flag = True\n",
    "        if seq_flag:\n",
    "            raise ValueError(\"Format incorrect, cannot trust result.\")\n",
    "        if len(seq) == 0:\n",
    "            raise ValueError(\"No data found, cannot trust result.\")\n",
    "    return seq\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# takes any genebank file (.gb) as input\n",
    "# returns a list of coding sequences (cds), where each item a string of the coding region range.\n",
    "def get_cds_positions(file):\n",
    "    with open(f\"{file}\", \"r\") as infile:\n",
    "        (cds_p, cds_flag) = (\"\", False)\n",
    "        for line in infile:\n",
    "            if line.startswith(\"     CDS             join\"):   # Green Flag\n",
    "                cds_flag = True\n",
    "            if cds_flag:\n",
    "                #.strip(\")\") to get rid of the extra \")\" at the end.\n",
    "                cds_p += line.strip(\"     CDS\").replace(\"join(\", \"\").replace(\")\", \"\").replace('/gene=\"COL1A2\"','').replace(\"\\n\",\"\")\n",
    "            if line.strip().startswith(\"/\"):    # Red Flag\n",
    "                cds_flag = False\n",
    "        if cds_flag:\n",
    "            raise ValueError(\"Format incorrect, cannot trust result.\")\n",
    "        if len(cds_p) == 0 :\n",
    "            raise ValueError(\"No data found, cannot trust result.\")\n",
    "    # cds is correctly retrieved\n",
    "    cds_p = cds_p.split(\",\")\n",
    "    return cds_p\n",
    "\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50ca1dca",
   "metadata": {},
   "source": [
    "4\\. This exercise builds on 2. You must read the dna7.fsa file and translate the DNA sequences to protein sequence. Report the frequencies of the various amino acids. That is count how many there is of each amino acid (a total) in the translated sequences, compute the frequency of each (Number_of_this_amino_acid/Total_number_of_amino_acids) and print the results as \"S 0.0123\", i.e. 4 digits after the dot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9d59444f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Below is a list of dictionaries, one for each dna sequence entry in dna7.fsa\n",
      "Each dictionary's key is an amino acid, and its value the frequency of its occurance in this once sequence entry\n",
      "[{'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}, {'M': '0.0420', 'A': '0.0840', 'L': '0.0588', 'W': '0.0084', 'T': '0.0504', 'R': '0.0420', 'V': '0.0504', 'P': '0.0756', 'H': '0.0168', 'F': '0.0672', 'N': '0.0420', 'Q': '0.0084', 'C': '0.0840', 'G': '0.0672', 'S': '0.0756', 'E': '0.0420', 'Y': '0.0252', 'K': '0.0504', 'I': '0.0420', 'D': '0.0672'}]\n"
     ]
    }
   ],
   "source": [
    "# pseudo code\n",
    "\"\"\"\n",
    "def main():\n",
    "    usr_aa_seqs = dna_to_aa(\"dna7.fsa\")\n",
    "    usr_aa_count = count_total(usr_aa_seqs)\n",
    "    \n",
    "    then initiate an empty dictionary: do things so that, if we:\n",
    "\n",
    "    have a dictionary which contains the amino acid as key, and the number of times the aa occured as value\n",
    "    returns a dictionary which contains amino acid as key and (Number_of_this_amino_acid/Total_number_of_amino_acids) as value\n",
    "    \n",
    "\n",
    "def count_total(aa_sequence):\n",
    "input a list where each item is an amino acid sequence\n",
    "outputs a dictionary where the key is the amino acid, the value is how many times in occured.\n",
    "\n",
    "\n",
    "def dna_to_aa(file):\n",
    "takes in a fsa file with multiple dna seq entries.\n",
    "outputs a list of protein sequences, where each item is a string.\n",
    "\n",
    "    with open the file\n",
    "        use pattern recognition to isolate the dna seq into a list of dna sequences.\n",
    "    for each dna sequence string:\n",
    "        read the dna sequence string three nucleotides at a time, translate it to a protein sequence string.\n",
    "    return a list, with each item being a protein sequence string.\n",
    "\n",
    "main()\n",
    "\"\"\"\n",
    "\n",
    "# takes in a list where each item is a string\n",
    "# and counting the occurences of each letter in the item\n",
    "# returns a dictionary where the keys are the letters,\n",
    "# and the value are the counts of the letters\n",
    "\n",
    "def main():\n",
    "    usr_aa_seqs = dna_to_aa(\"dna7.fsa\")\n",
    "    \n",
    "    letter_counted_list = []\n",
    "    for seq in usr_aa_seqs:\n",
    "        letter_counted_dict = dict()\n",
    "        for letter in seq:\n",
    "            if letter not in letter_counted_dict:\n",
    "                letter_counted_dict[letter] = 1\n",
    "            else:\n",
    "                letter_counted_dict[letter] += 1\n",
    "        letter_counted_list.append(letter_counted_dict)\n",
    "\n",
    "\n",
    "    letter_freq_list = []\n",
    "    letter_freq_dict = dict()\n",
    "    for dict_entry in letter_counted_list:\n",
    "        for k,v in dict_entry.items():\n",
    "            for seq in usr_aa_seqs:\n",
    "                freq = \"{:.4f}\".format(round(v/len(seq), 4))\n",
    "            letter_freq_dict[k] = freq\n",
    "        letter_freq_list.append(letter_freq_dict)\n",
    "    print(\"Below is a list of dictionaries, one for each dna sequence entry in dna7.fsa\")\n",
    "    print(\"Each dictionary's key is an amino acid, and its value the frequency of its occurance in this once sequence entry\")\n",
    "    print(letter_freq_list)\n",
    "\n",
    "\n",
    "def dna_to_aa(file):\n",
    "    dna_seqs=\"\"\n",
    "    with open(f\"{file}\", \"r\") as infile:\n",
    "        for line in infile:\n",
    "            if line.startswith(\">\"):\n",
    "                line = \">\"\n",
    "            else:\n",
    "                pass\n",
    "            # sanity check for replacement:\n",
    "            dna_seqs += line.strip()\n",
    "        # sanity check before \n",
    "    dna_seqs = dna_seqs.split(\">\")\n",
    "        # get rid of the empty string at the start.\n",
    "    dna_seqs.pop(0)\n",
    "        # passed sanity check :)\n",
    "    \n",
    "    # keys are codons:\n",
    "    codons_as_keys = list()\n",
    "    codons_as_keys = [(\"ATT\", \"ATC\", \"ATA\"), (\"CTT\", \"CTC\", \"CTA\", \"CTG\", \"TTA\", \"TTG\"), (\"GTT\", \"GTC\", \"GTA\", \"GTG\"), (\"TTT\", \"TTC\"), (\"ATG\"), (\"TGT\", \"TGC\"), (\"GCT\", \"GCC\", \"GCA\", \"GCG\"), (\"GGT\", \"GGC\", \"GGA\", \"GGG\"), (\"CCT\", \"CCC\", \"CCA\", \"CCG\"), (\"ACT\", \"ACC\", \"ACA\", \"ACG\"), (\"TCT\", \"TCC\", \"TCA\", \"TCG\", \"AGT\", \"AGC\"), (\"TAT\", \"TAC\"), (\"TGG\"), (\"CAA\", \"CAG\"), (\"AAT\", \"AAC\"), (\"CAT\", \"CAC\"), (\"GAA\", \"GAG\"), (\"GAT\", \"GAC\"), (\"AAA\", \"AAG\"), (\"CGT\", \"CGC\", \"CGA\", \"CGG\", \"AGA\", \"AGG\"), (\"TAA\", \"TAG\", \"TGA\")]\n",
    "    SLC_as_values = list()\n",
    "    SLC_as_values = [\"I\", \"L\", \"V\", \"F\", \"M\", \"C\", \"A\", \"G\", \"P\", \"T\", \"S\", \"Y\", \"W\", \"Q\", \"N\", \"H\", \"E\", \"D\", \"K\", \"R\", \"Stop\"]\n",
    "\n",
    "    codon_SLC_dict = dict(zip(codons_as_keys, SLC_as_values))\n",
    "\n",
    "    aa_seqs = []\n",
    "    for seq in dna_seqs:\n",
    "        aa_seq = \"\"\n",
    "        for position in range(0, len(seq)-3, 3):\n",
    "            codon = seq[position:position+3]\n",
    "            # have to use .item()\n",
    "            codon_SLC_list = codon_SLC_dict.items()\n",
    "            for k, v in codon_SLC_list:\n",
    "                if codon in k:\n",
    "                    aa_seq += v # no stop codons in the seqs\n",
    "        aa_seqs.append(aa_seq)\n",
    "    return aa_seqs\n",
    "\n",
    "\n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
