{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "88f8e77f",
   "metadata": {},
   "source": [
    "## L7: Stateful Parsing\n",
    "Exercises for extra practice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "726fec99",
   "metadata": {},
   "source": [
    "1\\. Count the number of RA (Author) lines in the sprot1-4.dat files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "df377b51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a sprot*.dat file name, for example, sprot2.dat: sprot2.dat\n",
      "25\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "\n",
    "def main():\n",
    "    file_to_work_on = input(\"Enter a sprot*.dat file name, for example, sprot2.dat: \")\n",
    "    RA_count(file_to_work_on)\n",
    "\n",
    "def RA_count(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ra_count, raflag) = (0, False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"RA\":\n",
    "            raflag = False\n",
    "        if line[:2] == \"RA\":\n",
    "            raflag = True\n",
    "        if raflag == True:\n",
    "            swissp_ra_count += 1\n",
    "    \n",
    "    if raflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "    \n",
    "    infile.close()\n",
    "    print(swissp_ra_count)\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4561fdf0",
   "metadata": {},
   "source": [
    "2\\. Extract the author names from the RA lines in the sprot1-4.dat files. Display the names - only the names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8fcf4707",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a sprot*.dat file name, for example, sprot2.dat: sprot2.dat\n",
      "Schuchman E.H. Suchi M. Takahashi T. Sandhoff K. Desnick R.J.\n",
      "Newrzella D. Stoffel W.\n",
      "Schuchman E.H. Levran O. Pereira L.V. Desnick R.J.\n",
      "Ida H. Rennert O.M. Eto Y. Chan W.Y.\n",
      "Quintern L.E. Schuchman E.H. Levran O. Suchi M. Ferlinz K. Reinke H. Sandhoff K. Desnick R.J.\n",
      "Ferlinz K. Hurwitz R. Moczall H. Lansmann S. Schuchman E.H. Sandhoff K.\n",
      "Lansmann S. Schuette C.G. Bartelsen O. Hoernschemeyer J. Linke T. Weisgerber J. Sandhoff K.\n",
      "Ferlinz K. Hurwitz R. Sandhoff K.\n",
      "Levran O. Desnick R.J. Schuchman E.H.\n",
      "Levran O. Desnick R.J. Schuchman E.H.\n",
      "Levran O. Desnick R.J. Schuchman E.H.\n",
      "Takahashi T. Desnick R.J. Takada G. Schuchman E.H.\n",
      "Takahashi T. Suchi M. Desnick R.J. Takada G. Schuchman E.H.\n",
      "Sperl W. Bart G. Vanier M.T. Christomanou H. Baldissera I. Steichensdorf E. Paschke E.\n",
      "Schuchman E.H.\n",
      "Takahashi T. Suchi M. Sato W. Ten S.B. Sakuragawa N. Desnick R.J. Schuchman E.H. Takada G.\n",
      "Ida H. Rennert O.M. Maekawa K. Eto Y.\n",
      "Simonaro C.M. Desnick R.J. McGovern M.M. Wasserstein M.P. Schuchman E.H.\n",
      "Sikora J. Pavlu-Pereira H. Elleder M. Roelofs H. Wevers R.A.\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "\n",
    "def main():\n",
    "    file_to_work_on = input(\"Enter a sprot*.dat file name, for example, sprot2.dat: \")\n",
    "    #ra_count(file_to_work_on)\n",
    "    read_print_ra(file_to_work_on)\n",
    "\n",
    "\n",
    "def ra_count(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ra_count, raflag) = (0, False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"RA\":\n",
    "            raflag = False\n",
    "        if line[:2] == \"RA\":\n",
    "            raflag = True\n",
    "        if raflag == True:\n",
    "            swissp_ra_count += 1\n",
    "    \n",
    "    if raflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "    \n",
    "    infile.close()\n",
    "    print(swissp_ra_count)\n",
    "\n",
    "# extract all authors into one line.\n",
    "def read_print_ra(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ra, raflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"RA\":\n",
    "            raflag = False\n",
    "        if line[:2] == \"RA\":\n",
    "            raflag = True\n",
    "        if raflag == True:\n",
    "            # adding the space add the end\n",
    "            # prevents authers in different lines from\n",
    "            # merging into the previous line\n",
    "            swissp_ra += line[2:].strip().replace(\",\", \"\") + \" \"\n",
    "    # first, all the authors from different papers are put into one big string.\n",
    "    # by replaying \";\" with \"\\n\", we correctly put the authors together\n",
    "    # i.e. authors of the same paper appear on the same line :)\n",
    "    swissp_ra = swissp_ra.replace(\";\", \"\\n\").replace(\"\\n \", \"\\n\")\n",
    "\n",
    "    if raflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "    \n",
    "    infile.close()\n",
    "    print(swissp_ra, end = \"\") #sanity check\n",
    "    with open(f\"{filename}_ra\", \"w\") as outfile:\n",
    "        outfile.write(swissp_ra)\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70c6bb99",
   "metadata": {},
   "source": [
    "3. Continuing previous exercise: Now also extract the title (RT lines). Display like title first on one line, followed by authors on the next. Then empty line followed by the next title and authors and so forth, until no more authors/titles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cbf564cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file name: sprot2.dat\n",
      "\"Human acid sphingomyelinase. Isolation nucleotide sequence and expression of the full-length and alternatively spliced cDNAs.\"\n",
      "Schuchman E.H. Suchi M. Takahashi T. Sandhoff K. Desnick R.J.\n",
      "\n",
      "\"Molecular cloning of the acid sphingomyelinase of the mouse and the organization and complete nucleotide sequence of the gene.\"\n",
      "Newrzella D. Stoffel W.\n",
      "\n",
      "\"Structural organization and complete nucleotide sequence of the gene encoding human acid sphingomyelinase (SMPD1).\"\n",
      "Schuchman E.H. Levran O. Pereira L.V. Desnick R.J.\n",
      "\n",
      "\"Cloning of a human acid sphingomyelinase cDNA with a new mutation that renders the enzyme inactive.\"\n",
      "Ida H. Rennert O.M. Eto Y. Chan W.Y.\n",
      "\n",
      "\"Isolation of cDNA clones encoding human acid sphingomyelinase: occurrence of alternatively processed transcripts.\"\n",
      "Quintern L.E. Schuchman E.H. Levran O. Suchi M. Ferlinz K. Reinke H. Sandhoff K. Desnick R.J.\n",
      "\n",
      "\"Functional characterization of the N-glycosylation sites of human acid sphingomyelinase by site-directed mutagenesis.\"\n",
      "Ferlinz K. Hurwitz R. Moczall H. Lansmann S. Schuchman E.H. Sandhoff K.\n",
      "\n",
      "\"Human acid sphingomyelinase.\"\n",
      "Lansmann S. Schuette C.G. Bartelsen O. Hoernschemeyer J. Linke T. Weisgerber J. Sandhoff K.\n",
      "\n",
      "\"Molecular basis of acid sphingomyelinase deficiency in a patient with Niemann-Pick disease type A.\"\n",
      "Ferlinz K. Hurwitz R. Sandhoff K.\n",
      "\n",
      "\"Niemann-Pick disease: a frequent missense mutation in the acid sphingomyelinase gene of Ashkenazi Jewish type A and B patients.\"\n",
      "Levran O. Desnick R.J. Schuchman E.H.\n",
      "\n",
      "\"Niemann-Pick type B disease. Identification of a single codon deletion in the acid sphingomyelinase gene and genotype/phenotype correlations in type A and B patients.\"\n",
      "Levran O. Desnick R.J. Schuchman E.H.\n",
      "\n",
      "\"Identification and expression of a common missense mutation (L302P) in the acid sphingomyelinase gene of Ashkenazi Jewish type A Niemann-Pick disease patients.\"\n",
      "Levran O. Desnick R.J. Schuchman E.H.\n",
      "\n",
      "\"Identification of a missense mutation (S436R) in the acid sphingomyelinase gene from a Japanese patient with type B Niemann-Pick disease.\"\n",
      "Takahashi T. Desnick R.J. Takada G. Schuchman E.H.\n",
      "\n",
      "\"Identification and expression of five mutations in the human acid sphingomyelinase gene causing types A and B Niemann-Pick disease. Molecular evidence for genetic heterogeneity in the neuronopathic and non-neuronopathic forms.\"\n",
      "Takahashi T. Suchi M. Desnick R.J. Takada G. Schuchman E.H.\n",
      "\n",
      "\"A family with visceral course of Niemann-Pick disease macular halo syndrome and low sphingomyelin degradation rate.\"\n",
      "Sperl W. Bart G. Vanier M.T. Christomanou H. Baldissera I. Steichensdorf E. Paschke E.\n",
      "\n",
      "\"Two new mutations in the acid sphingomyelinase gene causing type A Niemann-pick disease: N389T and R441X.\"\n",
      "Schuchman E.H.\n",
      "\n",
      "\"Identification and expression of a missense mutation (Y446C) in the acid sphingomyelinase gene from a Japanese patient with type A Niemann-Pick disease.\"\n",
      "Takahashi T. Suchi M. Sato W. Ten S.B. Sakuragawa N. Desnick R.J. Schuchman E.H. Takada G.\n",
      "\n",
      "\"Identification of three novel mutations in the acid sphingomyelinase gene of Japanese patients with Niemann-Pick disease type A and B.\"\n",
      "Ida H. Rennert O.M. Maekawa K. Eto Y.\n",
      "\n",
      "\"The demographics and distribution of type B Niemann-Pick disease: novel mutations lead to new genotype/phenotype correlations.\"\n",
      "Simonaro C.M. Desnick R.J. McGovern M.M. Wasserstein M.P. Schuchman E.H.\n",
      "\n",
      "\"Seven novel Acid sphingomyelinase gene mutations in Niemann-Pick type A and B patients.\"\n",
      "Sikora J. Pavlu-Pereira H. Elleder M. Roelofs H. Wevers R.A.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "\n",
    "def main():\n",
    "    file_to_work_on = input(\"Enter file name: \")\n",
    "    #ra_count(file_to_work_on)\n",
    "    read_ra(file_to_work_on)\n",
    "    read_rt(file_to_work_on)\n",
    "    display_rt_ra(file_to_work_on)\n",
    "\n",
    "\n",
    "def ra_count(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ra_count, raflag) = (0, False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"RA\":\n",
    "            raflag = False\n",
    "        if line[:2] == \"RA\":\n",
    "            raflag = True\n",
    "        if raflag == True:\n",
    "            swissp_ra_count += 1\n",
    "    \n",
    "    if raflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "    \n",
    "    infile.close()\n",
    "    print(swissp_ra_count)\n",
    "\n",
    "# extract all authors into one line.\n",
    "def read_ra(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_ra, raflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"RA\":\n",
    "            raflag = False\n",
    "        if line[:2] == \"RA\":\n",
    "            raflag = True\n",
    "        if raflag == True:\n",
    "            # adding the space add the end\n",
    "            # prevents authers in different lines from\n",
    "            # merging into the previous line\n",
    "            swissp_ra += line[2:].strip().replace(\",\", \"\") + \" \"\n",
    "    # first, all the authors from different papers are put into one big string.\n",
    "    # by replaying \";\" with \"\\n\", we correctly put the authors together\n",
    "    # i.e. authors of the same paper appear on the same line :)\n",
    "    swissp_ra = swissp_ra.replace(\";\", \"\\n\").replace(\"\\n \", \"\\n\")\n",
    "\n",
    "    if raflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "    \n",
    "    infile.close()\n",
    "#    print(swissp_ra, end = \"\") #sanity check\n",
    "    with open(f\"{filename}_ra\", \"w\") as outfile:\n",
    "        outfile.write(swissp_ra)\n",
    "\n",
    "# extract all authors into one line.\n",
    "def read_rt(filename):\n",
    "    try:\n",
    "        infile = open(f\"{filename}\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        sys.exit(1)\n",
    "    \n",
    "    (swissp_rt, rtflag) = (\"\", False)\n",
    "    for line in infile:\n",
    "        if line[:2] != \"RT\":\n",
    "            rtflag = False\n",
    "        if line[:2] == \"RT\":\n",
    "            rtflag = True\n",
    "        if rtflag == True:\n",
    "            # adding the space add the end\n",
    "            # prevents authers in different lines from\n",
    "            # merging into the previous line\n",
    "            swissp_rt += line[2:].strip().replace(\",\", \"\") + \" \"\n",
    "    # first, all the titles from different papers are put into one big string.\n",
    "    # by replaying \";\" with \"\\n\", we correctly separate each title.\n",
    "    swissp_rt = swissp_rt.replace(\";\", \"\\n\").replace(' \"', '\"')\n",
    "\n",
    "    if rtflag:\n",
    "        raise ValueError(\"Format is not right, can not trust result.\")\n",
    "    \n",
    "    infile.close()\n",
    "#    print()\n",
    "#    print(swissp_rt, end = \"\") #sanity check\n",
    "    with open(f\"{filename}_rt\", \"w\") as outfile:\n",
    "        outfile.write(swissp_rt)\n",
    "\n",
    "def display_rt_ra(filename):\n",
    "    try:\n",
    "        rt_file = open(f\"{filename}_rt\", \"r\")\n",
    "        ra_file = open(f\"{filename}_ra\", \"r\")\n",
    "    except IOError as err:\n",
    "        print(\"Can't open file, reason:\", str(err))\n",
    "        \n",
    "    rt_as_list = rt_file.readlines()\n",
    "    ra_as_list = ra_file.readlines()\n",
    "    ra_file.close()\n",
    "    rt_file.close()\n",
    "    rt_ra = \"\"\n",
    "    for i in range(len(ra_as_list)):    # len(rt_as_list) is 1 longer than that of ra.\n",
    "        rt_ra += rt_as_list[i] + ra_as_list[i] + \"\\n\"\n",
    "    print(rt_ra, end = \"\")\n",
    "\n",
    "\n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
